<!DOCTYPE html>

<html lang="en" xml:lang="en" xmlns="http://www.w3.org/1999/xhtml"><head>
<meta charset="utf-8"/>
<meta content="quarto-1.4.555" name="generator"/>
<meta content="width=device-width, initial-scale=1.0, user-scalable=yes" name="viewport"/>
<meta content="2024-06-11" name="dcterms.date"/>
<title>Aerobiome Integration in Air Quality measurements - Integrating air microbiome for comprehensive air quality analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>
<script crossorigin="anonymous" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta content="./" name="quarto:offset"/>
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet"/>
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" id="quarto-text-highlighting-styles" rel="stylesheet"/>
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet"/>
<link data-mode="light" href="site_libs/bootstrap/bootstrap.min.css" id="quarto-bootstrap" rel="stylesheet"/>
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>
<script crossorigin="anonymous" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
<link href="custom.css" rel="stylesheet"/>
</head>
<body class="nav-fixed">
<div id="quarto-search-results"></div>
<header class="headroom fixed-top" id="quarto-header">
<nav class="navbar navbar-expand-lg" data-bs-theme="dark">
<div class="navbar-container container-fluid">
<div class="navbar-brand-container mx-auto">
<a class="navbar-brand" href="./index.html">
<span class="navbar-title">Aerobiome Integration in Air Quality measurements</span>
</a>
</div>
<div class="" id="quarto-search" title="Search"></div>
<button aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" class="navbar-toggler" data-bs-target="#navbarCollapse" data-bs-toggle="collapse" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }" type="button">
<span class="navbar-toggler-icon"></span>
</button>
<div class="collapse navbar-collapse" id="navbarCollapse">
<ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
<a class="nav-link" href="./index.qmd">
<span class="menu-text">Home</span></a>
</li>
<li class="nav-item dropdown">
<a aria-expanded="false" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" href="#" id="nav-menu-about" role="button">
<span class="menu-text">About</span>
</a>
<ul aria-labelledby="nav-menu-about" class="dropdown-menu">
<li>
<a class="dropdown-item" href="./about.qmd">
<span class="dropdown-text">This Site</span></a>
</li>
<li>
<a class="dropdown-item" href="https://www.isglobal.org/en">
<span class="dropdown-text">ISGlobal</span></a>
</li>
<li>
<a class="dropdown-item" href="https://www.isglobal.org/en/-/clima-y-salud">
<span class="dropdown-text">Climate &amp; Health Programme</span></a>
</li>
<li>
<a class="dropdown-item" href="https://alfontal.dev">
<span class="dropdown-text">Alejandro Fontal</span></a>
</li>
</ul>
</li>
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item dropdown">
<a aria-expanded="false" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" href="#" id="nav-menu-bi-github" role="button">
<i class="bi bi-github" role="img">
</i>
<span class="menu-text"></span>
</a>
<ul aria-labelledby="nav-menu-bi-github" class="dropdown-menu dropdown-menu-end">
<li>
<a class="dropdown-item" href="https://github.com/AirLabBcn/aerobiome-integration-AQ">
<span class="dropdown-text">GitHub Repository</span></a>
</li>
<li>
<a class="dropdown-item" href="https://github.com/AirLabBcn/aerobiome-integration-AQ/issues/new">
<span class="dropdown-text">Raise an issue</span></a>
</li>
</ul>
</li>
<li class="nav-item compact">
<a class="nav-link" href="mailto:alejandro.fontal.92@gmail.com"> <i aria-label="Mail" class="bi bi-envelope" role="img">
</i>
<span class="menu-text"></span></a>
</li>
</ul>
</div> <!-- /navcollapse -->
<div class="quarto-navbar-tools">
</div>
</div> <!-- /container-fluid -->
</nav>
</header>
<!-- content -->
<div class="quarto-container page-columns page-rows-contents page-layout-article page-navbar" id="quarto-content">
<!-- sidebar -->
<!-- margin-sidebar -->
<div class="sidebar margin-sidebar" id="quarto-margin-sidebar">
<nav class="toc-active" id="TOC" role="doc-toc">
<h2 id="toc-title">CONTENTS</h2>
<ul>
<li><a class="nav-link active" data-scroll-target="#data-loading-and-wrangling" href="#data-loading-and-wrangling" id="toc-data-loading-and-wrangling">Data loading and wrangling</a>
<ul class="collapse">
<li><a class="nav-link" data-scroll-target="#experiment-1-continuous-vs-discrete-sampling" href="#experiment-1-continuous-vs-discrete-sampling" id="toc-experiment-1-continuous-vs-discrete-sampling">Experiment 1: Continuous vs Discrete sampling</a></li>
<li><a class="nav-link" data-scroll-target="#experiment-2-pm10-vs-pm2.5-heads" href="#experiment-2-pm10-vs-pm2.5-heads" id="toc-experiment-2-pm10-vs-pm2.5-heads">Experiment 2: PM10 vs PM2.5 heads</a></li>
<li><a class="nav-link" data-scroll-target="#supp.-experiment-bags-vs-vial-blanks" href="#supp.-experiment-bags-vs-vial-blanks" id="toc-supp.-experiment-bags-vs-vial-blanks">Supp. Experiment: Bags vs Vial blanks</a></li>
<li><a class="nav-link" data-scroll-target="#shannon-and-simpson-diversity-indexes" href="#shannon-and-simpson-diversity-indexes" id="toc-shannon-and-simpson-diversity-indexes">Shannon and Simpson diversity indexes</a></li>
</ul></li>
</ul>
</nav>
</div>
<!-- main -->
<main class="content" id="quarto-document-content">
<header class="quarto-title-block default" id="title-block-header">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Integrating air microbiome for comprehensive air quality analysis</h1><button aria-expanded="false" class="btn code-tools-button dropdown-toggle" data-bs-toggle="dropdown" id="quarto-code-tools-menu" type="button"><i class="bi"></i> Code</button><ul aria-labelelledby="quarto-code-tools-menu" class="dropdown-menu dropdown-menu-end"><li><a class="dropdown-item" href="javascript:void(0)" id="quarto-show-all-code" role="button">Show All Code</a></li><li><a class="dropdown-item" href="javascript:void(0)" id="quarto-hide-all-code" role="button">Hide All Code</a></li></ul></div></div>
<p class="subtitle lead">An evaluation of the potential of HVS for airborne microbiome monitoring</p>
</div>
<div class="quarto-title-meta-author">
<div class="quarto-title-meta-heading">Author</div>
<div class="quarto-title-meta-heading">Affiliation</div>
<div class="quarto-title-meta-contents">
<p class="author"><a href="https://alfontal.dev">Alejandro Fontal</a> <a class="quarto-title-author-email" href="mailto:alejandro.fontal.92@gmail.com"><i class="bi bi-envelope"></i></a> <a class="quarto-title-author-orcid" href="https://orcid.org/0000-0003-1138-2158"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="/></a></p>
</div>
<div class="quarto-title-meta-contents">
<p class="affiliation">
<a href="https://www.isglobal.org/en/-/clima-y-salud">
            AIRLAB, Climate and Health Programme, ISGlobal, Barcelona, Spain
            </a>
</p>
</div>
</div>
<div class="quarto-title-meta">
<div>
<div class="quarto-title-meta-heading">Published</div>
<div class="quarto-title-meta-contents">
<p class="date">June 11, 2024</p>
</div>
</div>
</div>
</header>
<p>In this report, we will perform part of the analysis and generate some of the figures for the manuscript <em>Integrating air microbiome for comprehensive air quality analysis</em>.</p>
<p>For easier reading, most code cells have been hidden by default, but you can expand them by clicking on the “Show code” button above each cell. The full code should be possible to run in a local environment top to bottom.</p>
<section class="level4 unnumbered" id="preamble">
<h4 class="unnumbered anchored" data-anchor-id="preamble">Preamble</h4>
<details>
<summary>
Show code imports and presets
</summary>
<p><br/> In this section we load the necessary libraries and set the global parameters for the notebook. It’s irrelevant for the narrative of the analysis but essential for the reproducibility of the results.</p>
<section class="level5 unnumbered" id="imports">
<h5 class="unnumbered anchored" data-anchor-id="imports">Imports</h5>
<div class="cell" data-execution_count="1" id="cell-6">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a aria-hidden="true" href="#cb1-1" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-2"><a aria-hidden="true" href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a aria-hidden="true" href="#cb1-3" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-4"><a aria-hidden="true" href="#cb1-4" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-5"><a aria-hidden="true" href="#cb1-5" tabindex="-1"></a><span class="im">import</span> plotnine <span class="im">as</span> p9</span>
<span id="cb1-6"><a aria-hidden="true" href="#cb1-6" tabindex="-1"></a><span class="im">import</span> matplotlib.colors</span>
<span id="cb1-7"><a aria-hidden="true" href="#cb1-7" tabindex="-1"></a><span class="im">import</span> scipy.stats <span class="im">as</span> stats</span>
<span id="cb1-8"><a aria-hidden="true" href="#cb1-8" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-9"><a aria-hidden="true" href="#cb1-9" tabindex="-1"></a></span>
<span id="cb1-10"><a aria-hidden="true" href="#cb1-10" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> Image</span>
<span id="cb1-11"><a aria-hidden="true" href="#cb1-11" tabindex="-1"></a><span class="im">from</span> mizani.formatters <span class="im">import</span> percent_format</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="level5 unnumbered" id="pre-sets">
<h5 class="unnumbered anchored" data-anchor-id="pre-sets">Pre-sets</h5>
<div class="cell" data-execution_count="2" id="cell-8">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a aria-hidden="true" href="#cb2-1" tabindex="-1"></a><span class="co"># Matplotlib settings</span></span>
<span id="cb2-2"><a aria-hidden="true" href="#cb2-2" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb2-3"><a aria-hidden="true" href="#cb2-3" tabindex="-1"></a><span class="im">from</span> matplotlib_inline.backend_inline <span class="im">import</span> set_matplotlib_formats</span>
<span id="cb2-4"><a aria-hidden="true" href="#cb2-4" tabindex="-1"></a>plt.rcParams[<span class="st">'font.family'</span>] <span class="op">=</span> <span class="st">'Georgia'</span></span>
<span id="cb2-5"><a aria-hidden="true" href="#cb2-5" tabindex="-1"></a>plt.rcParams[<span class="st">'svg.fonttype'</span>] <span class="op">=</span> <span class="st">'none'</span></span>
<span id="cb2-6"><a aria-hidden="true" href="#cb2-6" tabindex="-1"></a>set_matplotlib_formats(<span class="st">'retina'</span>)</span>
<span id="cb2-7"><a aria-hidden="true" href="#cb2-7" tabindex="-1"></a>plt.rcParams[<span class="st">'figure.dpi'</span>] <span class="op">=</span> <span class="dv">300</span></span>
<span id="cb2-8"><a aria-hidden="true" href="#cb2-8" tabindex="-1"></a></span>
<span id="cb2-9"><a aria-hidden="true" href="#cb2-9" tabindex="-1"></a><span class="co"># Plotnine settings (for figures)</span></span>
<span id="cb2-10"><a aria-hidden="true" href="#cb2-10" tabindex="-1"></a></span>
<span id="cb2-11"><a aria-hidden="true" href="#cb2-11" tabindex="-1"></a>p9.options.set_option(<span class="st">'base_family'</span>, <span class="st">'Georgia'</span>)</span>
<span id="cb2-12"><a aria-hidden="true" href="#cb2-12" tabindex="-1"></a></span>
<span id="cb2-13"><a aria-hidden="true" href="#cb2-13" tabindex="-1"></a>p9.theme_set(</span>
<span id="cb2-14"><a aria-hidden="true" href="#cb2-14" tabindex="-1"></a>    p9.theme_bw()</span>
<span id="cb2-15"><a aria-hidden="true" href="#cb2-15" tabindex="-1"></a>    <span class="op">+</span> p9.theme(panel_grid<span class="op">=</span>p9.element_blank(),</span>
<span id="cb2-16"><a aria-hidden="true" href="#cb2-16" tabindex="-1"></a>               legend_background<span class="op">=</span>p9.element_blank(),</span>
<span id="cb2-17"><a aria-hidden="true" href="#cb2-17" tabindex="-1"></a>               panel_grid_major<span class="op">=</span>p9.element_line(size<span class="op">=</span><span class="fl">.5</span>, linetype<span class="op">=</span><span class="st">'dashed'</span>,</span>
<span id="cb2-18"><a aria-hidden="true" href="#cb2-18" tabindex="-1"></a>                                                alpha<span class="op">=</span><span class="fl">.15</span>, color<span class="op">=</span><span class="st">'black'</span>),</span>
<span id="cb2-19"><a aria-hidden="true" href="#cb2-19" tabindex="-1"></a>               plot_title<span class="op">=</span>p9.element_text(ha<span class="op">=</span><span class="st">'center'</span>),</span>
<span id="cb2-20"><a aria-hidden="true" href="#cb2-20" tabindex="-1"></a>               dpi<span class="op">=</span><span class="dv">300</span></span>
<span id="cb2-21"><a aria-hidden="true" href="#cb2-21" tabindex="-1"></a>    )</span>
<span id="cb2-22"><a aria-hidden="true" href="#cb2-22" tabindex="-1"></a>)</span>
<span id="cb2-23"><a aria-hidden="true" href="#cb2-23" tabindex="-1"></a></span>
<span id="cb2-24"><a aria-hidden="true" href="#cb2-24" tabindex="-1"></a>separator_line <span class="op">=</span>  p9.annotate(</span>
<span id="cb2-25"><a aria-hidden="true" href="#cb2-25" tabindex="-1"></a>    geom<span class="op">=</span><span class="st">'line'</span>,</span>
<span id="cb2-26"><a aria-hidden="true" href="#cb2-26" tabindex="-1"></a>    linetype<span class="op">=</span><span class="st">'dashed'</span>,</span>
<span id="cb2-27"><a aria-hidden="true" href="#cb2-27" tabindex="-1"></a>    size<span class="op">=</span><span class="fl">.2</span>,</span>
<span id="cb2-28"><a aria-hidden="true" href="#cb2-28" tabindex="-1"></a>    x<span class="op">=</span>[<span class="dv">137</span>, <span class="dv">144</span>],</span>
<span id="cb2-29"><a aria-hidden="true" href="#cb2-29" tabindex="-1"></a>    y<span class="op">=</span> [<span class="dv">31</span>, <span class="dv">36</span>],</span>
<span id="cb2-30"><a aria-hidden="true" href="#cb2-30" tabindex="-1"></a>  )</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</div>
</section>
</details></section>
<section class="level2" id="data-loading-and-wrangling">
<h2 class="anchored" data-anchor-id="data-loading-and-wrangling">Data loading and wrangling</h2>
<p>We will start by loading the medatata file which we have stored in <code>/data/meta.txt</code> with information about each sample:</p>
<div class="cell" data-execution_count="3" id="cell-11">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a aria-hidden="true" href="#cb3-1" tabindex="-1"></a>metadata_df <span class="op">=</span> pd.read_table(<span class="st">'../data/meta.txt'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb3-2"><a aria-hidden="true" href="#cb3-2" tabindex="-1"></a>metadata_df.head()</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="3">
<div>
<table class="dataframe caption-top table table-sm table-striped small" data-border="1" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Sample Name</th>
<th data-quarto-table-cell-role="th">Sample Number</th>
<th data-quarto-table-cell-role="th">AirLab_Code</th>
<th data-quarto-table-cell-role="th">Seq ID</th>
<th data-quarto-table-cell-role="th">Sample ID</th>
<th data-quarto-table-cell-role="th">Sampling</th>
<th data-quarto-table-cell-role="th">Project Name</th>
<th data-quarto-table-cell-role="th">Location</th>
<th data-quarto-table-cell-role="th">Relative Location</th>
<th data-quarto-table-cell-role="th">Media Type</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Entered By</th>
<th data-quarto-table-cell-role="th">DNA Extraction Date (YYYY-MM-DD)</th>
<th data-quarto-table-cell-role="th">DNA Extraction Kit</th>
<th data-quarto-table-cell-role="th">DNA Concentration (ng/uL)</th>
<th data-quarto-table-cell-role="th">DNA Volume of Elution (uL)</th>
<th data-quarto-table-cell-role="th">DNA Quantification Instrument</th>
<th data-quarto-table-cell-role="th">DNA Vol. Used for Quantification (uL)</th>
<th data-quarto-table-cell-role="th">DNA Storage Freezer</th>
<th data-quarto-table-cell-role="th">DNA Storage Drawer</th>
<th data-quarto-table-cell-role="th">DNA Performed By</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>2022/11/23HVS-SASS-PilotA1</td>
<td>1</td>
<td>T6.0 B</td>
<td>HVSP1</td>
<td>C6-B</td>
<td>Continuous</td>
<td>HVS-SASS-Pilot</td>
<td>AirLab Terrace</td>
<td>Outdoor</td>
<td>Quartz Filter</td>
<td>...</td>
<td>Cassie Heinle</td>
<td>28/11/2022</td>
<td>Phe-Chlo method (AirLab Spain)</td>
<td>7133</td>
<td>17</td>
<td>Quantus Promega (AirLab Spain)</td>
<td>5</td>
<td>-20C MPL Freezer (New)</td>
<td>Yee Hui</td>
<td>AirLab Spain</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>2022/11/28HVS-SASS-PilotA2</td>
<td>2</td>
<td>T6.1 B</td>
<td>HVSP2</td>
<td>D6.1-B</td>
<td>Discrete</td>
<td>HVS-SASS-Pilot</td>
<td>AirLab Terrace</td>
<td>Outdoor</td>
<td>Quartz Filter</td>
<td>...</td>
<td>Cassie Heinle</td>
<td>28/11/2022</td>
<td>Phe-Chlo method (AirLab Spain)</td>
<td>856</td>
<td>17</td>
<td>Quantus Promega (AirLab Spain)</td>
<td>5</td>
<td>-20C MPL Freezer (New)</td>
<td>Yee Hui</td>
<td>AirLab Spain</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>2022/11/28HVS-SASS-PilotA3</td>
<td>3</td>
<td>T6.2 B</td>
<td>HVSP3</td>
<td>D6.2-B</td>
<td>Discrete</td>
<td>HVS-SASS-Pilot</td>
<td>AirLab Terrace</td>
<td>Outdoor</td>
<td>Quartz Filter</td>
<td>...</td>
<td>Cassie Heinle</td>
<td>28/11/2022</td>
<td>Phe-Chlo method (AirLab Spain)</td>
<td>1193</td>
<td>17</td>
<td>Quantus Promega (AirLab Spain)</td>
<td>5</td>
<td>-20C MPL Freezer (New)</td>
<td>Yee Hui</td>
<td>AirLab Spain</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>2022/11/28HVS-SASS-PilotA4</td>
<td>4</td>
<td>T6.3 B</td>
<td>HVSP4</td>
<td>D6.3-B</td>
<td>Discrete</td>
<td>HVS-SASS-Pilot</td>
<td>AirLab Terrace</td>
<td>Outdoor</td>
<td>Quartz Filter</td>
<td>...</td>
<td>Cassie Heinle</td>
<td>28/11/2022</td>
<td>Phe-Chlo method (AirLab Spain)</td>
<td>1483</td>
<td>17</td>
<td>Quantus Promega (AirLab Spain)</td>
<td>5</td>
<td>-20C MPL Freezer (New)</td>
<td>Yee Hui</td>
<td>AirLab Spain</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>2022/11/28HVS-SASS-PilotA5</td>
<td>5</td>
<td>T6.4 B</td>
<td>HVSP5</td>
<td>D6.4-B</td>
<td>Discrete</td>
<td>HVS-SASS-Pilot</td>
<td>AirLab Terrace</td>
<td>Outdoor</td>
<td>Quartz Filter</td>
<td>...</td>
<td>Cassie Heinle</td>
<td>28/11/2022</td>
<td>Phe-Chlo method (AirLab Spain)</td>
<td>111</td>
<td>17</td>
<td>Quantus Promega (AirLab Spain)</td>
<td>5</td>
<td>-20C MPL Freezer (New)</td>
<td>Yee Hui</td>
<td>AirLab Spain</td>
</tr>
</tbody>
</table>
<p>5 rows × 28 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="4" id="cell-12">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a aria-hidden="true" href="#cb4-1" tabindex="-1"></a>metadata_df <span class="op">=</span> (metadata_df</span>
<span id="cb4-2"><a aria-hidden="true" href="#cb4-2" tabindex="-1"></a> .drop(columns<span class="op">=</span>metadata_df.nunique().loc[<span class="kw">lambda</span> x: x<span class="op">==</span><span class="dv">1</span>].index)</span>
<span id="cb4-3"><a aria-hidden="true" href="#cb4-3" tabindex="-1"></a> .rename(columns<span class="op">=</span>{<span class="st">'Seq ID'</span>: <span class="st">'sample_id'</span>})</span>
<span id="cb4-4"><a aria-hidden="true" href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a aria-hidden="true" href="#cb4-5" tabindex="-1"></a>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</div>
<p>We also have the total number of reads for each assigned Species for all samples, which we will load from <code>/data/REL_hvsp_sp.txt</code></p>
<div class="cell" data-execution_count="5" id="cell-14">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a aria-hidden="true" href="#cb5-1" tabindex="-1"></a>species_df <span class="op">=</span> pd.read_table(<span class="st">'../data/REL_hvsp_sp.txt'</span>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</div>
<p>We have been given a list of species that are considered to be contaminants, which we will remove from the analysis:</p>
<div class="cell" data-execution_count="6" id="cell-16">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a aria-hidden="true" href="#cb6-1" tabindex="-1"></a>contamination_list <span class="op">=</span> [</span>
<span id="cb6-2"><a aria-hidden="true" href="#cb6-2" tabindex="-1"></a>    <span class="st">"Galendromus occidentalis"</span>, </span>
<span id="cb6-3"><a aria-hidden="true" href="#cb6-3" tabindex="-1"></a>    <span class="st">"Bemisia tabaci"</span>, </span>
<span id="cb6-4"><a aria-hidden="true" href="#cb6-4" tabindex="-1"></a>    <span class="st">"Candidatus Portiera aleyrodidarum"</span>, </span>
<span id="cb6-5"><a aria-hidden="true" href="#cb6-5" tabindex="-1"></a>    <span class="st">"Frankliniella occidentalis"</span>, </span>
<span id="cb6-6"><a aria-hidden="true" href="#cb6-6" tabindex="-1"></a>    <span class="st">"Thrips palmi"</span></span>
<span id="cb6-7"><a aria-hidden="true" href="#cb6-7" tabindex="-1"></a>]</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</div>
<p>Information on the total DNA yield of each sample is stored in the <code>/data/DNA.xlsx</code> file:</p>
<div class="cell" data-execution_count="7" id="cell-18">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a aria-hidden="true" href="#cb7-1" tabindex="-1"></a>dna_yields <span class="op">=</span> (pd.read_excel(<span class="st">'../data/DNA.xlsx'</span>)</span>
<span id="cb7-2"><a aria-hidden="true" href="#cb7-2" tabindex="-1"></a> .rename(columns<span class="op">=</span>{</span>
<span id="cb7-3"><a aria-hidden="true" href="#cb7-3" tabindex="-1"></a>    <span class="st">'DNA Concentration (ng/uL)'</span>: <span class="st">'dna_conc'</span>,</span>
<span id="cb7-4"><a aria-hidden="true" href="#cb7-4" tabindex="-1"></a>    <span class="st">'DNA Volume of Elution (uL)'</span>: <span class="st">'dna_vol'</span>,</span>
<span id="cb7-5"><a aria-hidden="true" href="#cb7-5" tabindex="-1"></a>    <span class="st">'Sample ID'</span>: <span class="st">'sample_code'</span></span>
<span id="cb7-6"><a aria-hidden="true" href="#cb7-6" tabindex="-1"></a> })</span>
<span id="cb7-7"><a aria-hidden="true" href="#cb7-7" tabindex="-1"></a> .assign(dna_yield<span class="op">=</span><span class="kw">lambda</span> x: x[<span class="st">'dna_conc'</span>] <span class="op">*</span> x[<span class="st">'dna_vol'</span>])</span>
<span id="cb7-8"><a aria-hidden="true" href="#cb7-8" tabindex="-1"></a> [[<span class="st">'sample_code'</span>, <span class="st">'dna_yield'</span>]]</span>
<span id="cb7-9"><a aria-hidden="true" href="#cb7-9" tabindex="-1"></a>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</div>
<p>We will generate a long-form dataframe with the species counts for each sample:</p>
<div class="cell" data-execution_count="8" id="cell-20">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a aria-hidden="true" href="#cb8-1" tabindex="-1"></a>species_long <span class="op">=</span> (species_df</span>
<span id="cb8-2"><a aria-hidden="true" href="#cb8-2" tabindex="-1"></a> .rename(columns<span class="op">=</span>{<span class="st">'#Datasets'</span>: <span class="st">'name'</span>})</span>
<span id="cb8-3"><a aria-hidden="true" href="#cb8-3" tabindex="-1"></a> .melt(<span class="st">'name'</span>, var_name<span class="op">=</span><span class="st">'sample_id'</span>, value_name<span class="op">=</span><span class="st">'reads'</span>)</span>
<span id="cb8-4"><a aria-hidden="true" href="#cb8-4" tabindex="-1"></a> .assign(sample_id<span class="op">=</span><span class="kw">lambda</span> x: x[<span class="st">'sample_id'</span>].<span class="bu">str</span>.split(<span class="st">'.'</span>).<span class="bu">str</span>[<span class="dv">0</span>])</span>
<span id="cb8-5"><a aria-hidden="true" href="#cb8-5" tabindex="-1"></a> .query(<span class="st">'name not in @contamination_list'</span>)</span>
<span id="cb8-6"><a aria-hidden="true" href="#cb8-6" tabindex="-1"></a> .query(<span class="st">'reads &gt; 0'</span>)</span>
<span id="cb8-7"><a aria-hidden="true" href="#cb8-7" tabindex="-1"></a>)</span>
<span id="cb8-8"><a aria-hidden="true" href="#cb8-8" tabindex="-1"></a>species_long.sample(<span class="dv">10</span>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<div>
<table class="dataframe caption-top table table-sm table-striped small" data-border="1" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">reads</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">128710</td>
<td>Krasilnikoviella flava</td>
<td>HVSP15</td>
<td>25.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">337027</td>
<td>Marmoricola sp. Leaf446</td>
<td>HVSP45</td>
<td>599.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">49047</td>
<td>Limibaculum sp. M0105</td>
<td>HVSP6</td>
<td>33.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">157388</td>
<td>Nocardioides ungokensis</td>
<td>HVSP19</td>
<td>148.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">269015</td>
<td>Rubripirellula reticaptiva</td>
<td>HVSP35</td>
<td>28.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">167215</td>
<td>Streptomyces hoynatensis</td>
<td>HVSP20</td>
<td>40.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">23868</td>
<td>Georgenia subflava</td>
<td>HVSP3</td>
<td>115.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">461309</td>
<td>Jeotgalicoccus psychrophilus</td>
<td>HVSP59</td>
<td>133.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">403987</td>
<td>Rubrobacter aplysinae</td>
<td>HVSP53</td>
<td>81.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">208014</td>
<td>Phyllostachys edulis</td>
<td>HVSP26</td>
<td>28.0</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<section class="level3" id="experiment-1-continuous-vs-discrete-sampling">
<h3 class="anchored" data-anchor-id="experiment-1-continuous-vs-discrete-sampling">Experiment 1: Continuous vs Discrete sampling</h3>
<p>Let’s start by taking a look at how the DNA yield of the samples and the diversity (measured as the richness of species) compares:</p>
<div class="cell" data-execution_count="9" id="cell-23">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a aria-hidden="true" href="#cb9-1" tabindex="-1"></a>exp_1_samples <span class="op">=</span> [<span class="st">'C3-B'</span>, <span class="st">'D3.1-B'</span>, <span class="st">'D3.2-B'</span>, <span class="st">'D3.3-B'</span>, <span class="st">'C6-B'</span>, <span class="st">'D6.1-B'</span>, <span class="st">'D6.2-B'</span>,</span>
<span id="cb9-2"><a aria-hidden="true" href="#cb9-2" tabindex="-1"></a>                  <span class="st">'D6.3-B'</span>, <span class="st">'D6.4-B'</span>, <span class="st">'D6.5-B'</span>, <span class="st">'D6.6-B'</span>,]</span>
<span id="cb9-3"><a aria-hidden="true" href="#cb9-3" tabindex="-1"></a>exp1_names <span class="op">=</span> [s.split(<span class="st">'-'</span>)[<span class="dv">0</span>] <span class="cf">for</span> s <span class="kw">in</span> exp_1_samples]</span>
<span id="cb9-4"><a aria-hidden="true" href="#cb9-4" tabindex="-1"></a>exp_1_meta <span class="op">=</span> (metadata_df</span>
<span id="cb9-5"><a aria-hidden="true" href="#cb9-5" tabindex="-1"></a> .query(<span class="st">'`Sample ID` in @exp_1_samples'</span>)</span>
<span id="cb9-6"><a aria-hidden="true" href="#cb9-6" tabindex="-1"></a> .assign(sampling_days<span class="op">=</span><span class="kw">lambda</span> x: x[<span class="st">'Sample ID'</span>].<span class="bu">str</span>[<span class="dv">1</span>])</span>
<span id="cb9-7"><a aria-hidden="true" href="#cb9-7" tabindex="-1"></a> .rename(columns<span class="op">=</span>{<span class="st">'Sample ID'</span>: <span class="st">'sample_code'</span>, <span class="st">'Sampling'</span>: <span class="st">'mode'</span>})</span>
<span id="cb9-8"><a aria-hidden="true" href="#cb9-8" tabindex="-1"></a> [[<span class="st">'sample_id'</span>, <span class="st">'sample_code'</span>, <span class="st">'mode'</span>, <span class="st">'sampling_days'</span>]]</span>
<span id="cb9-9"><a aria-hidden="true" href="#cb9-9" tabindex="-1"></a> .merge(dna_yields, on<span class="op">=</span><span class="st">'sample_code'</span>)</span>
<span id="cb9-10"><a aria-hidden="true" href="#cb9-10" tabindex="-1"></a>)</span>
<span id="cb9-11"><a aria-hidden="true" href="#cb9-11" tabindex="-1"></a></span>
<span id="cb9-12"><a aria-hidden="true" href="#cb9-12" tabindex="-1"></a>(species_long</span>
<span id="cb9-13"><a aria-hidden="true" href="#cb9-13" tabindex="-1"></a> .query(<span class="st">'reads &gt; 0'</span>)</span>
<span id="cb9-14"><a aria-hidden="true" href="#cb9-14" tabindex="-1"></a> .groupby(<span class="st">'sample_id'</span>)</span>
<span id="cb9-15"><a aria-hidden="true" href="#cb9-15" tabindex="-1"></a> .size()</span>
<span id="cb9-16"><a aria-hidden="true" href="#cb9-16" tabindex="-1"></a> .reset_index(name<span class="op">=</span><span class="st">'richness'</span>)</span>
<span id="cb9-17"><a aria-hidden="true" href="#cb9-17" tabindex="-1"></a> .merge(exp_1_meta)</span>
<span id="cb9-18"><a aria-hidden="true" href="#cb9-18" tabindex="-1"></a> .assign(sample_code<span class="op">=</span><span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.split(<span class="st">'-'</span>).<span class="bu">str</span>[<span class="dv">0</span>])</span>
<span id="cb9-19"><a aria-hidden="true" href="#cb9-19" tabindex="-1"></a> .assign(sample_code<span class="op">=</span><span class="kw">lambda</span> dd: </span>
<span id="cb9-20"><a aria-hidden="true" href="#cb9-20" tabindex="-1"></a>         pd.Categorical(dd.sample_code, categories<span class="op">=</span>exp1_names, ordered<span class="op">=</span><span class="va">True</span>))</span>
<span id="cb9-21"><a aria-hidden="true" href="#cb9-21" tabindex="-1"></a> .melt(id_vars<span class="op">=</span>[<span class="st">'sample_id'</span>, <span class="st">'sample_code'</span>, <span class="st">'mode'</span>, <span class="st">'sampling_days'</span>])</span>
<span id="cb9-22"><a aria-hidden="true" href="#cb9-22" tabindex="-1"></a> .replace({<span class="st">'richness'</span>: <span class="st">'Species richness'</span>, <span class="st">'dna_yield'</span>: <span class="st">'DNA yield (ng)'</span>})</span>
<span id="cb9-23"><a aria-hidden="true" href="#cb9-23" tabindex="-1"></a> .pipe(<span class="kw">lambda</span> x: p9.ggplot(x) </span>
<span id="cb9-24"><a aria-hidden="true" href="#cb9-24" tabindex="-1"></a>       <span class="op">+</span> p9.aes(x<span class="op">=</span><span class="st">'sample_code'</span>, y<span class="op">=</span><span class="st">'value'</span>, fill<span class="op">=</span><span class="st">'mode'</span>)</span>
<span id="cb9-25"><a aria-hidden="true" href="#cb9-25" tabindex="-1"></a>       <span class="op">+</span> p9.geom_col()</span>
<span id="cb9-26"><a aria-hidden="true" href="#cb9-26" tabindex="-1"></a>       <span class="op">+</span> p9.facet_wrap(<span class="st">'~variable'</span>, scales<span class="op">=</span><span class="st">'free_y'</span>, ncol<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb9-27"><a aria-hidden="true" href="#cb9-27" tabindex="-1"></a>       <span class="op">+</span> p9.annotate(<span class="st">'vline'</span>, xintercept<span class="op">=</span>[<span class="fl">4.5</span>], linetype<span class="op">=</span><span class="st">'dashed'</span>, size<span class="op">=</span><span class="fl">.4</span>)</span>
<span id="cb9-28"><a aria-hidden="true" href="#cb9-28" tabindex="-1"></a>       <span class="op">+</span> p9.labs(x<span class="op">=</span><span class="st">''</span>, y<span class="op">=</span><span class="st">''</span>, fill<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb9-29"><a aria-hidden="true" href="#cb9-29" tabindex="-1"></a>       <span class="op">+</span> p9.theme(</span>
<span id="cb9-30"><a aria-hidden="true" href="#cb9-30" tabindex="-1"></a>           figure_size<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">4</span>),</span>
<span id="cb9-31"><a aria-hidden="true" href="#cb9-31" tabindex="-1"></a>           axis_text_x<span class="op">=</span>p9.element_text(angle<span class="op">=</span><span class="dv">45</span>),</span>
<span id="cb9-32"><a aria-hidden="true" href="#cb9-32" tabindex="-1"></a>           legend_position<span class="op">=</span><span class="st">'top'</span>,</span>
<span id="cb9-33"><a aria-hidden="true" href="#cb9-33" tabindex="-1"></a>           legend_key_size<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb9-34"><a aria-hidden="true" href="#cb9-34" tabindex="-1"></a>           legend_text<span class="op">=</span>p9.element_text(size<span class="op">=</span><span class="dv">7</span>),</span>
<span id="cb9-35"><a aria-hidden="true" href="#cb9-35" tabindex="-1"></a>           )</span>
<span id="cb9-36"><a aria-hidden="true" href="#cb9-36" tabindex="-1"></a>       )</span>
<span id="cb9-37"><a aria-hidden="true" href="#cb9-37" tabindex="-1"></a>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img class="img-fluid figure-img" src="HVS-SASS_files/figure-html/cell-10-output-1.png"/></p>
</figure>
</div>
</div>
</div>
<p>At a simple glance, it seems that the continuous sampling method yields more DNA (which makes sense since they are sampling for longer periods of time) but this does not seem to translate into a higher species richness.</p>
<p>If we now check not only the richness but the actual species belonging to each of the sample types and see whether some species are more common in one type of sample than the other:</p>
<div class="cell" data-execution_count="10" id="cell-26">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a aria-hidden="true" href="#cb10-1" tabindex="-1"></a>table_3 <span class="op">=</span> (species_long</span>
<span id="cb10-2"><a aria-hidden="true" href="#cb10-2" tabindex="-1"></a> .merge(exp_1_meta)</span>
<span id="cb10-3"><a aria-hidden="true" href="#cb10-3" tabindex="-1"></a> .query(<span class="st">'sampling_days.notna()'</span>)</span>
<span id="cb10-4"><a aria-hidden="true" href="#cb10-4" tabindex="-1"></a> .query(<span class="st">'reads &gt; 0'</span>)</span>
<span id="cb10-5"><a aria-hidden="true" href="#cb10-5" tabindex="-1"></a> .query(<span class="st">'sampling_days == "3"'</span>)</span>
<span id="cb10-6"><a aria-hidden="true" href="#cb10-6" tabindex="-1"></a> .groupby([<span class="st">'name'</span>, <span class="st">'mode'</span>, <span class="st">'sampling_days'</span>])</span>
<span id="cb10-7"><a aria-hidden="true" href="#cb10-7" tabindex="-1"></a> [<span class="st">'reads'</span>]</span>
<span id="cb10-8"><a aria-hidden="true" href="#cb10-8" tabindex="-1"></a> .<span class="bu">sum</span>()</span>
<span id="cb10-9"><a aria-hidden="true" href="#cb10-9" tabindex="-1"></a> .reset_index()</span>
<span id="cb10-10"><a aria-hidden="true" href="#cb10-10" tabindex="-1"></a> .pivot(index<span class="op">=</span><span class="st">'name'</span>, columns<span class="op">=</span>[<span class="st">'mode'</span>], values<span class="op">=</span><span class="st">'reads'</span>)</span>
<span id="cb10-11"><a aria-hidden="true" href="#cb10-11" tabindex="-1"></a> .fillna(<span class="dv">0</span>)</span>
<span id="cb10-12"><a aria-hidden="true" href="#cb10-12" tabindex="-1"></a> .loc[<span class="kw">lambda</span> x: x.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>) <span class="op">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb10-13"><a aria-hidden="true" href="#cb10-13" tabindex="-1"></a> .applymap(<span class="kw">lambda</span> x: x <span class="op">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb10-14"><a aria-hidden="true" href="#cb10-14" tabindex="-1"></a> .assign(presence<span class="op">=</span><span class="kw">lambda</span> dd: </span>
<span id="cb10-15"><a aria-hidden="true" href="#cb10-15" tabindex="-1"></a>         np.where(dd[<span class="st">'Continuous'</span>] <span class="op">==</span> dd[<span class="st">'Discrete'</span>], <span class="st">'Both'</span>,</span>
<span id="cb10-16"><a aria-hidden="true" href="#cb10-16" tabindex="-1"></a>         np.where(dd[<span class="st">'Continuous'</span>], <span class="st">'Only Continuous'</span>, <span class="st">'Only Discrete'</span>))</span>
<span id="cb10-17"><a aria-hidden="true" href="#cb10-17" tabindex="-1"></a>         )</span>
<span id="cb10-18"><a aria-hidden="true" href="#cb10-18" tabindex="-1"></a> .dropna()</span>
<span id="cb10-19"><a aria-hidden="true" href="#cb10-19" tabindex="-1"></a> .groupby([<span class="st">'presence'</span>])</span>
<span id="cb10-20"><a aria-hidden="true" href="#cb10-20" tabindex="-1"></a> .size()</span>
<span id="cb10-21"><a aria-hidden="true" href="#cb10-21" tabindex="-1"></a> .reset_index()</span>
<span id="cb10-22"><a aria-hidden="true" href="#cb10-22" tabindex="-1"></a> .rename(columns<span class="op">=</span>{<span class="dv">0</span>:<span class="st">'richness'</span>})</span>
<span id="cb10-23"><a aria-hidden="true" href="#cb10-23" tabindex="-1"></a> .assign(freq<span class="op">=</span><span class="kw">lambda</span> dd: dd[<span class="st">'richness'</span>] <span class="op">/</span> dd[<span class="st">'richness'</span>].<span class="bu">sum</span>())</span>
<span id="cb10-24"><a aria-hidden="true" href="#cb10-24" tabindex="-1"></a> .assign(days<span class="op">=</span><span class="st">"3 days"</span>)</span>
<span id="cb10-25"><a aria-hidden="true" href="#cb10-25" tabindex="-1"></a> )</span>
<span id="cb10-26"><a aria-hidden="true" href="#cb10-26" tabindex="-1"></a></span>
<span id="cb10-27"><a aria-hidden="true" href="#cb10-27" tabindex="-1"></a>table_6 <span class="op">=</span> (species_long</span>
<span id="cb10-28"><a aria-hidden="true" href="#cb10-28" tabindex="-1"></a> .merge(exp_1_meta)</span>
<span id="cb10-29"><a aria-hidden="true" href="#cb10-29" tabindex="-1"></a> .query(<span class="st">'sampling_days.notna()'</span>)</span>
<span id="cb10-30"><a aria-hidden="true" href="#cb10-30" tabindex="-1"></a> .query(<span class="st">'reads &gt; 0'</span>)</span>
<span id="cb10-31"><a aria-hidden="true" href="#cb10-31" tabindex="-1"></a> .query(<span class="st">'sampling_days == "6"'</span>)</span>
<span id="cb10-32"><a aria-hidden="true" href="#cb10-32" tabindex="-1"></a> .groupby([<span class="st">'name'</span>, <span class="st">'mode'</span>, <span class="st">'sampling_days'</span>])</span>
<span id="cb10-33"><a aria-hidden="true" href="#cb10-33" tabindex="-1"></a> [<span class="st">'reads'</span>]</span>
<span id="cb10-34"><a aria-hidden="true" href="#cb10-34" tabindex="-1"></a> .<span class="bu">sum</span>()</span>
<span id="cb10-35"><a aria-hidden="true" href="#cb10-35" tabindex="-1"></a> .reset_index()</span>
<span id="cb10-36"><a aria-hidden="true" href="#cb10-36" tabindex="-1"></a> .pivot(index<span class="op">=</span><span class="st">'name'</span>, columns<span class="op">=</span>[<span class="st">'mode'</span>], values<span class="op">=</span><span class="st">'reads'</span>)</span>
<span id="cb10-37"><a aria-hidden="true" href="#cb10-37" tabindex="-1"></a> .fillna(<span class="dv">0</span>)</span>
<span id="cb10-38"><a aria-hidden="true" href="#cb10-38" tabindex="-1"></a> .loc[<span class="kw">lambda</span> x: x.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>) <span class="op">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb10-39"><a aria-hidden="true" href="#cb10-39" tabindex="-1"></a> .applymap(<span class="kw">lambda</span> x: x <span class="op">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb10-40"><a aria-hidden="true" href="#cb10-40" tabindex="-1"></a> .assign(presence<span class="op">=</span><span class="kw">lambda</span> dd: </span>
<span id="cb10-41"><a aria-hidden="true" href="#cb10-41" tabindex="-1"></a>         np.where(dd[<span class="st">'Continuous'</span>] <span class="op">==</span> dd[<span class="st">'Discrete'</span>], <span class="st">'Both'</span>,</span>
<span id="cb10-42"><a aria-hidden="true" href="#cb10-42" tabindex="-1"></a>         np.where(dd[<span class="st">'Continuous'</span>], <span class="st">'Only Continuous'</span>, <span class="st">'Only Discrete'</span>))</span>
<span id="cb10-43"><a aria-hidden="true" href="#cb10-43" tabindex="-1"></a>         )</span>
<span id="cb10-44"><a aria-hidden="true" href="#cb10-44" tabindex="-1"></a> .dropna()</span>
<span id="cb10-45"><a aria-hidden="true" href="#cb10-45" tabindex="-1"></a> .groupby([<span class="st">'presence'</span>])</span>
<span id="cb10-46"><a aria-hidden="true" href="#cb10-46" tabindex="-1"></a> .size()</span>
<span id="cb10-47"><a aria-hidden="true" href="#cb10-47" tabindex="-1"></a> .reset_index()</span>
<span id="cb10-48"><a aria-hidden="true" href="#cb10-48" tabindex="-1"></a> .rename(columns<span class="op">=</span>{<span class="dv">0</span>:<span class="st">'richness'</span>})</span>
<span id="cb10-49"><a aria-hidden="true" href="#cb10-49" tabindex="-1"></a> .assign(freq<span class="op">=</span><span class="kw">lambda</span> dd: dd[<span class="st">'richness'</span>] <span class="op">/</span> dd[<span class="st">'richness'</span>].<span class="bu">sum</span>())</span>
<span id="cb10-50"><a aria-hidden="true" href="#cb10-50" tabindex="-1"></a> .assign(days<span class="op">=</span><span class="st">"6 days"</span>)</span>
<span id="cb10-51"><a aria-hidden="true" href="#cb10-51" tabindex="-1"></a> )</span>
<span id="cb10-52"><a aria-hidden="true" href="#cb10-52" tabindex="-1"></a></span>
<span id="cb10-53"><a aria-hidden="true" href="#cb10-53" tabindex="-1"></a></span>
<span id="cb10-54"><a aria-hidden="true" href="#cb10-54" tabindex="-1"></a>(pd.concat([table_3, table_6])</span>
<span id="cb10-55"><a aria-hidden="true" href="#cb10-55" tabindex="-1"></a> .pipe(<span class="kw">lambda</span> dd: p9.ggplot(dd)</span>
<span id="cb10-56"><a aria-hidden="true" href="#cb10-56" tabindex="-1"></a>       <span class="op">+</span> p9.aes(<span class="st">'days'</span>, <span class="st">'freq'</span>)</span>
<span id="cb10-57"><a aria-hidden="true" href="#cb10-57" tabindex="-1"></a>       <span class="op">+</span> p9.geom_col(p9.aes(fill<span class="op">=</span><span class="st">'presence'</span>), width<span class="op">=</span><span class="fl">.5</span>)</span>
<span id="cb10-58"><a aria-hidden="true" href="#cb10-58" tabindex="-1"></a>       <span class="op">+</span> p9.scale_y_continuous(labels<span class="op">=</span>percent_format())</span>
<span id="cb10-59"><a aria-hidden="true" href="#cb10-59" tabindex="-1"></a>       <span class="op">+</span> p9.scale_fill_manual([<span class="st">'#4B9C78'</span>, <span class="st">'#D85E01'</span>, <span class="st">'#7B76B5'</span></span>
<span id="cb10-60"><a aria-hidden="true" href="#cb10-60" tabindex="-1"></a>           </span>
<span id="cb10-61"><a aria-hidden="true" href="#cb10-61" tabindex="-1"></a>       ])</span>
<span id="cb10-62"><a aria-hidden="true" href="#cb10-62" tabindex="-1"></a>       <span class="op">+</span> p9.labs(x<span class="op">=</span><span class="st">'Sampling Duration'</span>, y<span class="op">=</span><span class="st">'Percentage of unique species'</span>, fill<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb10-63"><a aria-hidden="true" href="#cb10-63" tabindex="-1"></a>       <span class="op">+</span> p9.theme(figure_size<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">3</span>),</span>
<span id="cb10-64"><a aria-hidden="true" href="#cb10-64" tabindex="-1"></a>                  legend_position<span class="op">=</span><span class="st">'top'</span></span>
<span id="cb10-65"><a aria-hidden="true" href="#cb10-65" tabindex="-1"></a>                  )</span>
<span id="cb10-66"><a aria-hidden="true" href="#cb10-66" tabindex="-1"></a></span>
<span id="cb10-67"><a aria-hidden="true" href="#cb10-67" tabindex="-1"></a></span>
<span id="cb10-68"><a aria-hidden="true" href="#cb10-68" tabindex="-1"></a> )</span>
<span id="cb10-69"><a aria-hidden="true" href="#cb10-69" tabindex="-1"></a>       )</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img class="img-fluid figure-img" src="HVS-SASS_files/figure-html/cell-11-output-1.png"/></p>
</figure>
</div>
</div>
</div>
<p>If we represent the same information in a Sankey diagram:</p>
<div class="cell" data-execution_count="11" id="cell-28">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a aria-hidden="true" href="#cb11-1" tabindex="-1"></a>Image(<span class="st">'../output/sankey_3d_vs_6d.png'</span>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="11">
<div>
<figure class="figure">
<p><img class="img-fluid figure-img" src="HVS-SASS_files/figure-html/cell-12-output-1.png"/></p>
</figure>
</div>
</div>
</div>
</section>
<section class="level3" id="experiment-2-pm10-vs-pm2.5-heads">
<h3 class="anchored" data-anchor-id="experiment-2-pm10-vs-pm2.5-heads">Experiment 2: PM10 vs PM2.5 heads</h3>
<p>If we take a look now at the 3 daily samples taken with a PM10 head and the 3 daily samples taken with a PM2.5 head, and where the species identified appear:</p>
<div class="cell" data-execution_count="12" id="cell-31">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a aria-hidden="true" href="#cb12-1" tabindex="-1"></a>(species_long</span>
<span id="cb12-2"><a aria-hidden="true" href="#cb12-2" tabindex="-1"></a>.merge(metadata_df[[<span class="st">'Sampling'</span>, <span class="st">'Sample ID'</span>, <span class="st">'sample_id'</span>]])</span>
<span id="cb12-3"><a aria-hidden="true" href="#cb12-3" tabindex="-1"></a>.rename(columns<span class="op">=</span>{<span class="st">'Sample ID'</span>: <span class="st">'sample_code'</span>})</span>
<span id="cb12-4"><a aria-hidden="true" href="#cb12-4" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'PM'</span>)]</span>
<span id="cb12-5"><a aria-hidden="true" href="#cb12-5" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.endswith(<span class="st">'-B'</span>)]</span>
<span id="cb12-6"><a aria-hidden="true" href="#cb12-6" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'.1.'</span>, regex<span class="op">=</span><span class="va">False</span>)]</span>
<span id="cb12-7"><a aria-hidden="true" href="#cb12-7" tabindex="-1"></a>.assign(header<span class="op">=</span><span class="kw">lambda</span> dd: np.where(dd.sample_code.<span class="bu">str</span>.contains(<span class="st">"2.5"</span>, regex<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb12-8"><a aria-hidden="true" href="#cb12-8" tabindex="-1"></a>                                   <span class="st">'PM2.5'</span>, <span class="st">'PM10'</span></span>
<span id="cb12-9"><a aria-hidden="true" href="#cb12-9" tabindex="-1"></a>                                   ))</span>
<span id="cb12-10"><a aria-hidden="true" href="#cb12-10" tabindex="-1"></a>.query(<span class="st">'reads &gt; 0'</span>)</span>
<span id="cb12-11"><a aria-hidden="true" href="#cb12-11" tabindex="-1"></a>.pivot_table(index<span class="op">=</span>[<span class="st">'name'</span>], columns<span class="op">=</span><span class="st">'header'</span>, values<span class="op">=</span><span class="st">'reads'</span>, aggfunc<span class="op">=</span><span class="st">'sum'</span>)</span>
<span id="cb12-12"><a aria-hidden="true" href="#cb12-12" tabindex="-1"></a>.assign(presence<span class="op">=</span><span class="kw">lambda</span> dd: np.where(dd[<span class="st">'PM10'</span>].isna(), <span class="st">'PM2.5'</span>,</span>
<span id="cb12-13"><a aria-hidden="true" href="#cb12-13" tabindex="-1"></a>                            np.where(dd[<span class="st">'PM2.5'</span>].isna(), <span class="st">'PM10'</span>, <span class="st">'Both'</span>)))</span>
<span id="cb12-14"><a aria-hidden="true" href="#cb12-14" tabindex="-1"></a>.presence.value_counts()</span>
<span id="cb12-15"><a aria-hidden="true" href="#cb12-15" tabindex="-1"></a>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>presence
Both     3634
PM10      841
PM2.5     740
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>We see that in this case, from a total of 5215 species, 3634 appear in both types of samples, 841 are unique to PM10 samples and 740 are unique to PM2.5 samples.</p>
<p>I am going to now generate a table listing the species unique to each sample type and the count of unique samples where they appear. The formatted tables can be accessed <a href="https://docs.google.com/spreadsheets/d/1Z2mpG_UcF1rGa4fqV-FreewOr0SFzgNETQ_teDca9Qc/edit?usp=sharing">here</a>.</p>
<div class="cell" data-execution_count="13" id="cell-34">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a aria-hidden="true" href="#cb14-1" tabindex="-1"></a>uniquely_unique <span class="op">=</span> (species_long</span>
<span id="cb14-2"><a aria-hidden="true" href="#cb14-2" tabindex="-1"></a>.merge(metadata_df[[<span class="st">'Sampling'</span>, <span class="st">'Sample ID'</span>, <span class="st">'sample_id'</span>]])</span>
<span id="cb14-3"><a aria-hidden="true" href="#cb14-3" tabindex="-1"></a>.rename(columns<span class="op">=</span>{<span class="st">'Sample ID'</span>: <span class="st">'sample_code'</span>})</span>
<span id="cb14-4"><a aria-hidden="true" href="#cb14-4" tabindex="-1"></a>.query(<span class="st">'name not in @contamination_list'</span>)</span>
<span id="cb14-5"><a aria-hidden="true" href="#cb14-5" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'PM'</span>)]</span>
<span id="cb14-6"><a aria-hidden="true" href="#cb14-6" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.endswith(<span class="st">'-B'</span>)]</span>
<span id="cb14-7"><a aria-hidden="true" href="#cb14-7" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'.1.'</span>, regex<span class="op">=</span><span class="va">False</span>)]</span>
<span id="cb14-8"><a aria-hidden="true" href="#cb14-8" tabindex="-1"></a>.assign(header<span class="op">=</span><span class="kw">lambda</span> dd: np.where(dd.sample_code.<span class="bu">str</span>.contains(<span class="st">"2.5"</span>, regex<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb14-9"><a aria-hidden="true" href="#cb14-9" tabindex="-1"></a>                                   <span class="st">'PM2.5'</span>, <span class="st">'PM10'</span></span>
<span id="cb14-10"><a aria-hidden="true" href="#cb14-10" tabindex="-1"></a>                                   ))</span>
<span id="cb14-11"><a aria-hidden="true" href="#cb14-11" tabindex="-1"></a>.query(<span class="st">'reads &gt; 0'</span>)</span>
<span id="cb14-12"><a aria-hidden="true" href="#cb14-12" tabindex="-1"></a>.groupby([<span class="st">'name'</span>, <span class="st">'header'</span>])</span>
<span id="cb14-13"><a aria-hidden="true" href="#cb14-13" tabindex="-1"></a>.size()</span>
<span id="cb14-14"><a aria-hidden="true" href="#cb14-14" tabindex="-1"></a>.rename(<span class="st">'n_present'</span>)</span>
<span id="cb14-15"><a aria-hidden="true" href="#cb14-15" tabindex="-1"></a>.reset_index()</span>
<span id="cb14-16"><a aria-hidden="true" href="#cb14-16" tabindex="-1"></a>.pivot(index<span class="op">=</span><span class="st">'name'</span>, columns<span class="op">=</span><span class="st">'header'</span>, values<span class="op">=</span><span class="st">'n_present'</span>)</span>
<span id="cb14-17"><a aria-hidden="true" href="#cb14-17" tabindex="-1"></a>.fillna(<span class="dv">0</span>)</span>
<span id="cb14-18"><a aria-hidden="true" href="#cb14-18" tabindex="-1"></a>.astype(<span class="bu">int</span>)</span>
<span id="cb14-19"><a aria-hidden="true" href="#cb14-19" tabindex="-1"></a>.query(<span class="st">'`PM2.5`==0 or `PM10`==0'</span>)</span>
<span id="cb14-20"><a aria-hidden="true" href="#cb14-20" tabindex="-1"></a>.query(<span class="st">'`PM2.5`&gt;=1 or `PM10`&gt;=1'</span>)</span>
<span id="cb14-21"><a aria-hidden="true" href="#cb14-21" tabindex="-1"></a>.index</span>
<span id="cb14-22"><a aria-hidden="true" href="#cb14-22" tabindex="-1"></a>.values</span>
<span id="cb14-23"><a aria-hidden="true" href="#cb14-23" tabindex="-1"></a>)</span>
<span id="cb14-24"><a aria-hidden="true" href="#cb14-24" tabindex="-1"></a></span>
<span id="cb14-25"><a aria-hidden="true" href="#cb14-25" tabindex="-1"></a>unique_pm_reads <span class="op">=</span> (species_long</span>
<span id="cb14-26"><a aria-hidden="true" href="#cb14-26" tabindex="-1"></a>.merge(metadata_df[[<span class="st">'Sampling'</span>, <span class="st">'Sample ID'</span>, <span class="st">'sample_id'</span>]])</span>
<span id="cb14-27"><a aria-hidden="true" href="#cb14-27" tabindex="-1"></a>.rename(columns<span class="op">=</span>{<span class="st">'Sample ID'</span>: <span class="st">'sample_code'</span>})</span>
<span id="cb14-28"><a aria-hidden="true" href="#cb14-28" tabindex="-1"></a>.query(<span class="st">'name not in @contamination_list'</span>)</span>
<span id="cb14-29"><a aria-hidden="true" href="#cb14-29" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'PM'</span>)]</span>
<span id="cb14-30"><a aria-hidden="true" href="#cb14-30" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.endswith(<span class="st">'-B'</span>)]</span>
<span id="cb14-31"><a aria-hidden="true" href="#cb14-31" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'.1.'</span>, regex<span class="op">=</span><span class="va">False</span>)]</span>
<span id="cb14-32"><a aria-hidden="true" href="#cb14-32" tabindex="-1"></a>.assign(header<span class="op">=</span><span class="kw">lambda</span> dd: np.where(dd.sample_code.<span class="bu">str</span>.contains(<span class="st">"2.5"</span>, regex<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb14-33"><a aria-hidden="true" href="#cb14-33" tabindex="-1"></a>                                   <span class="st">'PM2.5'</span>, <span class="st">'PM10'</span></span>
<span id="cb14-34"><a aria-hidden="true" href="#cb14-34" tabindex="-1"></a>                                   ))</span>
<span id="cb14-35"><a aria-hidden="true" href="#cb14-35" tabindex="-1"></a>.query(<span class="st">'reads &gt; 0'</span>)</span>
<span id="cb14-36"><a aria-hidden="true" href="#cb14-36" tabindex="-1"></a>.query(<span class="st">'name in @uniquely_unique'</span>)</span>
<span id="cb14-37"><a aria-hidden="true" href="#cb14-37" tabindex="-1"></a>.groupby([<span class="st">'name'</span>, <span class="st">'header'</span>])</span>
<span id="cb14-38"><a aria-hidden="true" href="#cb14-38" tabindex="-1"></a>.agg({<span class="st">'reads'</span>: <span class="bu">sum</span>, <span class="st">'sample_code'</span>: <span class="st">'nunique'</span>})</span>
<span id="cb14-39"><a aria-hidden="true" href="#cb14-39" tabindex="-1"></a>.reset_index()</span>
<span id="cb14-40"><a aria-hidden="true" href="#cb14-40" tabindex="-1"></a>)</span>
<span id="cb14-41"><a aria-hidden="true" href="#cb14-41" tabindex="-1"></a></span>
<span id="cb14-42"><a aria-hidden="true" href="#cb14-42" tabindex="-1"></a>(unique_pm_reads</span>
<span id="cb14-43"><a aria-hidden="true" href="#cb14-43" tabindex="-1"></a> .pivot(index<span class="op">=</span><span class="st">'name'</span>, columns<span class="op">=</span><span class="st">'header'</span>)</span>
<span id="cb14-44"><a aria-hidden="true" href="#cb14-44" tabindex="-1"></a> .fillna(<span class="dv">0</span>)</span>
<span id="cb14-45"><a aria-hidden="true" href="#cb14-45" tabindex="-1"></a> .astype(<span class="bu">int</span>)</span>
<span id="cb14-46"><a aria-hidden="true" href="#cb14-46" tabindex="-1"></a>).to_csv(<span class="st">'../output/table_unique_species_pm.csv'</span>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>If we check the total reads assigned per sample day and head type, we observe that while for days 1 and 2 the total reads where rather high and similar for both head types, but the 3rd day had a strong reduction, with the PM2.5 head having a much lower number of reads than the PM10 head (which still had a low number of reads compared to the previous two days).</p>
<div class="cell" data-execution_count="14" id="cell-36">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a aria-hidden="true" href="#cb15-1" tabindex="-1"></a>(species_long</span>
<span id="cb15-2"><a aria-hidden="true" href="#cb15-2" tabindex="-1"></a>.merge(metadata_df[[<span class="st">'Sampling'</span>, <span class="st">'Sample ID'</span>, <span class="st">'sample_id'</span>]])</span>
<span id="cb15-3"><a aria-hidden="true" href="#cb15-3" tabindex="-1"></a>.rename(columns<span class="op">=</span>{<span class="st">'Sample ID'</span>: <span class="st">'sample_code'</span>})</span>
<span id="cb15-4"><a aria-hidden="true" href="#cb15-4" tabindex="-1"></a>.query(<span class="st">'name not in @contamination_list'</span>)</span>
<span id="cb15-5"><a aria-hidden="true" href="#cb15-5" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'PM'</span>)]</span>
<span id="cb15-6"><a aria-hidden="true" href="#cb15-6" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.endswith(<span class="st">'-B'</span>)]</span>
<span id="cb15-7"><a aria-hidden="true" href="#cb15-7" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'.1.'</span>, regex<span class="op">=</span><span class="va">False</span>)]</span>
<span id="cb15-8"><a aria-hidden="true" href="#cb15-8" tabindex="-1"></a>.assign(header<span class="op">=</span><span class="kw">lambda</span> dd: np.where(dd.sample_code.<span class="bu">str</span>.contains(<span class="st">"2.5"</span>, regex<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb15-9"><a aria-hidden="true" href="#cb15-9" tabindex="-1"></a>                                   <span class="st">'PM2.5'</span>, <span class="st">'PM10'</span></span>
<span id="cb15-10"><a aria-hidden="true" href="#cb15-10" tabindex="-1"></a>                                   ))</span>
<span id="cb15-11"><a aria-hidden="true" href="#cb15-11" tabindex="-1"></a>.query(<span class="st">'reads &gt; 0'</span>)</span>
<span id="cb15-12"><a aria-hidden="true" href="#cb15-12" tabindex="-1"></a>.groupby([<span class="st">'header'</span>, <span class="st">'sample_code'</span>])</span>
<span id="cb15-13"><a aria-hidden="true" href="#cb15-13" tabindex="-1"></a>.reads.<span class="bu">sum</span>()</span>
<span id="cb15-14"><a aria-hidden="true" href="#cb15-14" tabindex="-1"></a>.reset_index()</span>
<span id="cb15-15"><a aria-hidden="true" href="#cb15-15" tabindex="-1"></a>.assign(sample_code<span class="op">=</span><span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.split(<span class="st">'-'</span>).<span class="bu">str</span>[<span class="dv">0</span>])</span>
<span id="cb15-16"><a aria-hidden="true" href="#cb15-16" tabindex="-1"></a>.assign(day<span class="op">=</span><span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.split(<span class="st">'.'</span>).<span class="bu">str</span>[<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb15-17"><a aria-hidden="true" href="#cb15-17" tabindex="-1"></a>.pipe(<span class="kw">lambda</span> dd: p9.ggplot(dd)</span>
<span id="cb15-18"><a aria-hidden="true" href="#cb15-18" tabindex="-1"></a>      <span class="op">+</span> p9.aes(<span class="st">'day'</span>, <span class="st">'reads'</span>, fill<span class="op">=</span><span class="st">'header'</span>)</span>
<span id="cb15-19"><a aria-hidden="true" href="#cb15-19" tabindex="-1"></a>      <span class="op">+</span> p9.geom_col(position<span class="op">=</span><span class="st">'dodge'</span>)</span>
<span id="cb15-20"><a aria-hidden="true" href="#cb15-20" tabindex="-1"></a>      <span class="op">+</span> p9.coord_flip()</span>
<span id="cb15-21"><a aria-hidden="true" href="#cb15-21" tabindex="-1"></a>      <span class="op">+</span> p9.scale_fill_manual([<span class="st">'#D85E01'</span>, <span class="st">'#7B76B5'</span>])</span>
<span id="cb15-22"><a aria-hidden="true" href="#cb15-22" tabindex="-1"></a>      <span class="op">+</span> p9.labs(x<span class="op">=</span><span class="st">'Discrete Sampling Day'</span>, y<span class="op">=</span><span class="st">'Total Reads'</span>, fill<span class="op">=</span><span class="st">'Sampling head'</span>)</span>
<span id="cb15-23"><a aria-hidden="true" href="#cb15-23" tabindex="-1"></a>      <span class="op">+</span> p9.theme(</span>
<span id="cb15-24"><a aria-hidden="true" href="#cb15-24" tabindex="-1"></a>          figure_size<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">3</span>),</span>
<span id="cb15-25"><a aria-hidden="true" href="#cb15-25" tabindex="-1"></a>          legend_position<span class="op">=</span>(<span class="fl">.8</span>, <span class="fl">.85</span>),</span>
<span id="cb15-26"><a aria-hidden="true" href="#cb15-26" tabindex="-1"></a>          legend_key_size<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb15-27"><a aria-hidden="true" href="#cb15-27" tabindex="-1"></a>          legend_title<span class="op">=</span>p9.element_text(x<span class="op">=</span><span class="dv">80</span>, y<span class="op">=</span><span class="dv">50</span>, size<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb15-28"><a aria-hidden="true" href="#cb15-28" tabindex="-1"></a>          )</span>
<span id="cb15-29"><a aria-hidden="true" href="#cb15-29" tabindex="-1"></a>      )</span>
<span id="cb15-30"><a aria-hidden="true" href="#cb15-30" tabindex="-1"></a>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img class="img-fluid figure-img" src="HVS-SASS_files/figure-html/cell-15-output-1.png"/></p>
</figure>
</div>
</div>
</div>
<p>If we focus on the species which are not only unique to each head type but that are present with higher relative weight in one of the head types, we can see the top 50 most differentiallly abundant species here (based on the diffs between the relative abundance of the species in the two head types):</p>
<div class="cell" data-execution_count="17" id="cell-38">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a aria-hidden="true" href="#cb16-1" tabindex="-1"></a>species_by_header <span class="op">=</span> (species_long</span>
<span id="cb16-2"><a aria-hidden="true" href="#cb16-2" tabindex="-1"></a>.merge(metadata_df[[<span class="st">'Sampling'</span>, <span class="st">'Sample ID'</span>, <span class="st">'sample_id'</span>]])</span>
<span id="cb16-3"><a aria-hidden="true" href="#cb16-3" tabindex="-1"></a>.rename(columns<span class="op">=</span>{<span class="st">'Sample ID'</span>: <span class="st">'sample_code'</span>})</span>
<span id="cb16-4"><a aria-hidden="true" href="#cb16-4" tabindex="-1"></a>.query(<span class="st">'name not in @contamination_list'</span>)</span>
<span id="cb16-5"><a aria-hidden="true" href="#cb16-5" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'PM'</span>)]</span>
<span id="cb16-6"><a aria-hidden="true" href="#cb16-6" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.endswith(<span class="st">'-B'</span>)]</span>
<span id="cb16-7"><a aria-hidden="true" href="#cb16-7" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'.1.'</span>, regex<span class="op">=</span><span class="va">False</span>)]</span>
<span id="cb16-8"><a aria-hidden="true" href="#cb16-8" tabindex="-1"></a>.assign(header<span class="op">=</span><span class="kw">lambda</span> dd: np.where(dd.sample_code.<span class="bu">str</span>.contains(<span class="st">"2.5"</span>, regex<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb16-9"><a aria-hidden="true" href="#cb16-9" tabindex="-1"></a>                                   <span class="st">'PM2.5'</span>, <span class="st">'PM10'</span></span>
<span id="cb16-10"><a aria-hidden="true" href="#cb16-10" tabindex="-1"></a>                                   ))</span>
<span id="cb16-11"><a aria-hidden="true" href="#cb16-11" tabindex="-1"></a>.groupby([<span class="st">'header'</span>, <span class="st">'sample_code'</span>], as_index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-12"><a aria-hidden="true" href="#cb16-12" tabindex="-1"></a>.<span class="bu">apply</span>(<span class="kw">lambda</span> dd: dd.assign(rel_abundance<span class="op">=</span><span class="kw">lambda</span> x: x[<span class="st">'reads'</span>] <span class="op">/</span> x[<span class="st">'reads'</span>].<span class="bu">sum</span>()))</span>
<span id="cb16-13"><a aria-hidden="true" href="#cb16-13" tabindex="-1"></a>.reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb16-14"><a aria-hidden="true" href="#cb16-14" tabindex="-1"></a>.groupby([<span class="st">'name'</span>, <span class="st">'header'</span>])</span>
<span id="cb16-15"><a aria-hidden="true" href="#cb16-15" tabindex="-1"></a>.agg({<span class="st">'rel_abundance'</span>: <span class="kw">lambda</span> x: x.<span class="bu">sum</span>() <span class="op">/</span> <span class="dv">3</span>, <span class="st">'reads'</span>: <span class="st">'sum'</span>})</span>
<span id="cb16-16"><a aria-hidden="true" href="#cb16-16" tabindex="-1"></a>.loc[<span class="kw">lambda</span> x: x.reads <span class="op">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb16-17"><a aria-hidden="true" href="#cb16-17" tabindex="-1"></a>.reset_index()</span>
<span id="cb16-18"><a aria-hidden="true" href="#cb16-18" tabindex="-1"></a>.pivot(index<span class="op">=</span><span class="st">'name'</span>, columns<span class="op">=</span><span class="st">'header'</span>, values<span class="op">=</span>[<span class="st">'rel_abundance'</span>, <span class="st">'reads'</span>])</span>
<span id="cb16-19"><a aria-hidden="true" href="#cb16-19" tabindex="-1"></a>.fillna(<span class="dv">0</span>)</span>
<span id="cb16-20"><a aria-hidden="true" href="#cb16-20" tabindex="-1"></a>.assign(diff<span class="op">=</span><span class="kw">lambda</span> dd: (dd[(<span class="st">'rel_abundance'</span>,  <span class="st">'PM2.5'</span>)] <span class="op">-</span> dd[(<span class="st">'rel_abundance'</span>,  <span class="st">'PM10'</span>)]))</span>
<span id="cb16-21"><a aria-hidden="true" href="#cb16-21" tabindex="-1"></a>.assign(direction<span class="op">=</span><span class="kw">lambda</span> dd: np.where(dd[<span class="st">'diff'</span>] <span class="op">&gt;</span> <span class="dv">0</span>, <span class="st">'PM2.5 &gt; PM10'</span>, <span class="st">'PM2.5 &lt; PM10'</span>))</span>
<span id="cb16-22"><a aria-hidden="true" href="#cb16-22" tabindex="-1"></a>.assign(abs_diff<span class="op">=</span><span class="kw">lambda</span> dd: dd[<span class="st">'diff'</span>].<span class="bu">abs</span>())</span>
<span id="cb16-23"><a aria-hidden="true" href="#cb16-23" tabindex="-1"></a>.sort_values(<span class="st">'diff'</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb16-24"><a aria-hidden="true" href="#cb16-24" tabindex="-1"></a>.head(<span class="dv">50</span>)</span>
<span id="cb16-25"><a aria-hidden="true" href="#cb16-25" tabindex="-1"></a>.assign(reads_label<span class="op">=</span><span class="kw">lambda</span> dd: </span>
<span id="cb16-26"><a aria-hidden="true" href="#cb16-26" tabindex="-1"></a>        dd[<span class="st">'reads'</span>, <span class="st">'PM10'</span>].astype(<span class="bu">int</span>).astype(<span class="bu">str</span>) <span class="op">+</span> <span class="st">':'</span> <span class="op">+</span> </span>
<span id="cb16-27"><a aria-hidden="true" href="#cb16-27" tabindex="-1"></a>        dd[<span class="st">'reads'</span>, <span class="st">'PM2.5'</span>].astype(<span class="bu">int</span>).astype(<span class="bu">str</span>))</span>
<span id="cb16-28"><a aria-hidden="true" href="#cb16-28" tabindex="-1"></a>.reset_index()</span>
<span id="cb16-29"><a aria-hidden="true" href="#cb16-29" tabindex="-1"></a>.assign(direction<span class="op">=</span><span class="kw">lambda</span> dd: </span>
<span id="cb16-30"><a aria-hidden="true" href="#cb16-30" tabindex="-1"></a>        pd.Categorical(dd.direction, categories<span class="op">=</span>[<span class="st">'PM2.5 &gt; PM10'</span>, <span class="st">'PM2.5 &lt; PM10'</span>],</span>
<span id="cb16-31"><a aria-hidden="true" href="#cb16-31" tabindex="-1"></a>                       ordered<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb16-32"><a aria-hidden="true" href="#cb16-32" tabindex="-1"></a>)</span>
<span id="cb16-33"><a aria-hidden="true" href="#cb16-33" tabindex="-1"></a> .replace({<span class="st">"Candidatus Symbiopectobacterium sp. 'North America'"</span>:</span>
<span id="cb16-34"><a aria-hidden="true" href="#cb16-34" tabindex="-1"></a>           <span class="st">"Candidatus Symbiopectobacterium sp."</span>})</span>
<span id="cb16-35"><a aria-hidden="true" href="#cb16-35" tabindex="-1"></a>)</span>
<span id="cb16-36"><a aria-hidden="true" href="#cb16-36" tabindex="-1"></a></span>
<span id="cb16-37"><a aria-hidden="true" href="#cb16-37" tabindex="-1"></a>sorted_species_rel <span class="op">=</span> species_by_header.sort_values((<span class="st">'rel_abundance'</span>, <span class="st">'PM2.5'</span>))[<span class="st">'name'</span>]</span>
<span id="cb16-38"><a aria-hidden="true" href="#cb16-38" tabindex="-1"></a>sorted_species_reads <span class="op">=</span> species_by_header.sort_values((<span class="st">'reads'</span>, <span class="st">'PM10'</span>))[<span class="st">'name'</span>]</span>
<span id="cb16-39"><a aria-hidden="true" href="#cb16-39" tabindex="-1"></a></span>
<span id="cb16-40"><a aria-hidden="true" href="#cb16-40" tabindex="-1"></a>species_by_header.columns <span class="op">=</span> [<span class="st">'_'</span>.join(col).strip() <span class="cf">for</span> col <span class="kw">in</span> species_by_header.columns.values]</span>
<span id="cb16-41"><a aria-hidden="true" href="#cb16-41" tabindex="-1"></a>(species_by_header</span>
<span id="cb16-42"><a aria-hidden="true" href="#cb16-42" tabindex="-1"></a> .assign(rel_abundance_PM10<span class="op">=</span><span class="kw">lambda</span> dd: dd[<span class="st">'rel_abundance_PM10'</span>] <span class="op">*</span> <span class="op">-</span> <span class="dv">1</span>)</span>
<span id="cb16-43"><a aria-hidden="true" href="#cb16-43" tabindex="-1"></a> .assign(label_pm25<span class="op">=</span><span class="kw">lambda</span> dd: dd[<span class="st">'reads_PM2.5'</span>].astype(<span class="bu">int</span>).astype(<span class="bu">str</span>) <span class="op">+</span>  <span class="st">' ('</span></span>
<span id="cb16-44"><a aria-hidden="true" href="#cb16-44" tabindex="-1"></a>         <span class="op">+</span> (dd[<span class="st">'rel_abundance_PM2.5'</span>] <span class="op">*</span> <span class="dv">100</span>).<span class="bu">round</span>(<span class="dv">2</span>).astype(<span class="bu">str</span>) <span class="op">+</span> <span class="st">'%)'</span>)</span>
<span id="cb16-45"><a aria-hidden="true" href="#cb16-45" tabindex="-1"></a> .assign(label_pm10<span class="op">=</span><span class="kw">lambda</span> dd: dd[<span class="st">'reads_PM10'</span>].astype(<span class="bu">int</span>).astype(<span class="bu">str</span>) <span class="op">+</span>  <span class="st">' ('</span></span>
<span id="cb16-46"><a aria-hidden="true" href="#cb16-46" tabindex="-1"></a>         <span class="op">+</span> (dd[<span class="st">'rel_abundance_PM10'</span>].<span class="bu">abs</span>() <span class="op">*</span> <span class="dv">100</span>).<span class="bu">round</span>(<span class="dv">2</span>).astype(<span class="bu">str</span>) <span class="op">+</span> <span class="st">'%)'</span>)</span>
<span id="cb16-47"><a aria-hidden="true" href="#cb16-47" tabindex="-1"></a>.assign(name<span class="op">=</span><span class="kw">lambda</span> dd: pd.Categorical(dd.name_, categories<span class="op">=</span>sorted_species_rel, ordered<span class="op">=</span><span class="va">True</span>))</span>
<span id="cb16-48"><a aria-hidden="true" href="#cb16-48" tabindex="-1"></a>.pipe(<span class="kw">lambda</span> dd: p9.ggplot(dd) </span>
<span id="cb16-49"><a aria-hidden="true" href="#cb16-49" tabindex="-1"></a>      <span class="op">+</span> p9.aes(<span class="st">'name'</span>)</span>
<span id="cb16-50"><a aria-hidden="true" href="#cb16-50" tabindex="-1"></a>      <span class="op">+</span> p9.coord_flip() </span>
<span id="cb16-51"><a aria-hidden="true" href="#cb16-51" tabindex="-1"></a>      <span class="op">+</span> p9.geom_col(p9.aes(y<span class="op">=</span><span class="st">'rel_abundance_PM2.5'</span>), fill<span class="op">=</span><span class="st">'#7B76B5'</span>, width<span class="op">=</span><span class="fl">.6</span>)</span>
<span id="cb16-52"><a aria-hidden="true" href="#cb16-52" tabindex="-1"></a>      <span class="op">+</span> p9.geom_col(p9.aes(y<span class="op">=</span><span class="st">'rel_abundance_PM10'</span>), fill<span class="op">=</span><span class="st">'#D85E01'</span>, width<span class="op">=</span><span class="fl">.6</span>)</span>
<span id="cb16-53"><a aria-hidden="true" href="#cb16-53" tabindex="-1"></a>      <span class="op">+</span> p9.geom_text(p9.aes(label<span class="op">=</span><span class="st">'label_pm10'</span>, y<span class="op">=</span><span class="st">'rel_abundance_PM10'</span>),</span>
<span id="cb16-54"><a aria-hidden="true" href="#cb16-54" tabindex="-1"></a>                      size<span class="op">=</span><span class="dv">5</span>, ha<span class="op">=</span><span class="st">'right'</span>, nudge_y<span class="op">=-</span><span class="fl">.0005</span>)</span>
<span id="cb16-55"><a aria-hidden="true" href="#cb16-55" tabindex="-1"></a>      <span class="op">+</span> p9.geom_text(p9.aes(label<span class="op">=</span><span class="st">'label_pm25'</span>, y<span class="op">=</span><span class="st">'rel_abundance_PM2.5'</span>),</span>
<span id="cb16-56"><a aria-hidden="true" href="#cb16-56" tabindex="-1"></a>                      size<span class="op">=</span><span class="dv">5</span>, ha<span class="op">=</span><span class="st">'left'</span>, nudge_y<span class="op">=</span><span class="fl">.0005</span>)</span>
<span id="cb16-57"><a aria-hidden="true" href="#cb16-57" tabindex="-1"></a>      <span class="op">+</span> p9.scale_y_continuous(</span>
<span id="cb16-58"><a aria-hidden="true" href="#cb16-58" tabindex="-1"></a>          labels<span class="op">=</span>[<span class="st">"2%"</span>, <span class="st">"0%"</span>, <span class="st">"2%"</span>],</span>
<span id="cb16-59"><a aria-hidden="true" href="#cb16-59" tabindex="-1"></a>          limits<span class="op">=</span>[<span class="op">-</span><span class="fl">.034</span>, <span class="fl">.034</span>])</span>
<span id="cb16-60"><a aria-hidden="true" href="#cb16-60" tabindex="-1"></a>      <span class="op">+</span> p9.annotate(<span class="st">'hline'</span>, yintercept<span class="op">=</span><span class="dv">0</span>, linetype<span class="op">=</span><span class="st">'solid'</span>, size<span class="op">=</span><span class="fl">.2</span>)</span>
<span id="cb16-61"><a aria-hidden="true" href="#cb16-61" tabindex="-1"></a>      <span class="op">+</span> p9.labs(x<span class="op">=</span><span class="st">''</span>, y<span class="op">=</span><span class="st">'Mean Relative Abundance'</span>, title<span class="op">=</span><span class="ss">f'PM10              PM2.5'</span>)</span>
<span id="cb16-62"><a aria-hidden="true" href="#cb16-62" tabindex="-1"></a>      <span class="op">+</span> p9.theme(figure_size<span class="op">=</span>(<span class="fl">4.5</span>, <span class="dv">5</span>),</span>
<span id="cb16-63"><a aria-hidden="true" href="#cb16-63" tabindex="-1"></a>                 axis_text_y<span class="op">=</span>p9.element_text(size<span class="op">=</span><span class="fl">6.5</span>),</span>
<span id="cb16-64"><a aria-hidden="true" href="#cb16-64" tabindex="-1"></a>                 axis_title_x<span class="op">=</span>p9.element_text(size<span class="op">=</span><span class="dv">9</span>),</span>
<span id="cb16-65"><a aria-hidden="true" href="#cb16-65" tabindex="-1"></a>                 plot_title<span class="op">=</span>p9.element_text(size<span class="op">=</span><span class="dv">10</span>, ha<span class="op">=</span><span class="st">'center'</span>),</span>
<span id="cb16-66"><a aria-hidden="true" href="#cb16-66" tabindex="-1"></a>                 )</span>
<span id="cb16-67"><a aria-hidden="true" href="#cb16-67" tabindex="-1"></a>      )</span>
<span id="cb16-68"><a aria-hidden="true" href="#cb16-68" tabindex="-1"></a>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img class="img-fluid figure-img" src="HVS-SASS_files/figure-html/cell-16-output-1.png"/></p>
</figure>
</div>
</div>
</div>
<p>If we instead focus on the species that appear in at least 2 samples of one head type and none of the other, the top 30 for each head type are:</p>
<div class="cell" data-execution_count="19" id="cell-40">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a aria-hidden="true" href="#cb17-1" tabindex="-1"></a>f <span class="op">=</span> (unique_pm_reads</span>
<span id="cb17-2"><a aria-hidden="true" href="#cb17-2" tabindex="-1"></a>     .query(<span class="st">'sample_code &gt;= 2'</span>)</span>
<span id="cb17-3"><a aria-hidden="true" href="#cb17-3" tabindex="-1"></a> .groupby(<span class="st">'header'</span>)</span>
<span id="cb17-4"><a aria-hidden="true" href="#cb17-4" tabindex="-1"></a> .<span class="bu">apply</span>(<span class="kw">lambda</span> dd: dd</span>
<span id="cb17-5"><a aria-hidden="true" href="#cb17-5" tabindex="-1"></a>        .sort_values(<span class="st">'reads'</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb17-6"><a aria-hidden="true" href="#cb17-6" tabindex="-1"></a>        .reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb17-7"><a aria-hidden="true" href="#cb17-7" tabindex="-1"></a>        .assign(rank<span class="op">=</span><span class="kw">lambda</span> x: x.index <span class="op">+</span> <span class="dv">1</span>)</span>
<span id="cb17-8"><a aria-hidden="true" href="#cb17-8" tabindex="-1"></a>        )</span>
<span id="cb17-9"><a aria-hidden="true" href="#cb17-9" tabindex="-1"></a> .reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb17-10"><a aria-hidden="true" href="#cb17-10" tabindex="-1"></a> .query(<span class="st">'rank &lt;= 30'</span>)</span>
<span id="cb17-11"><a aria-hidden="true" href="#cb17-11" tabindex="-1"></a> .replace({<span class="st">'PM10'</span>: <span class="st">'Present only in PM10'</span>, <span class="st">'PM2.5'</span>: <span class="st">'Present only in PM2.5'</span>})</span>
<span id="cb17-12"><a aria-hidden="true" href="#cb17-12" tabindex="-1"></a> .pipe(<span class="kw">lambda</span> dd: p9.ggplot(dd)</span>
<span id="cb17-13"><a aria-hidden="true" href="#cb17-13" tabindex="-1"></a>       <span class="op">+</span> p9.aes(<span class="st">'reorder(name, reads)'</span>, <span class="st">'reads'</span>)</span>
<span id="cb17-14"><a aria-hidden="true" href="#cb17-14" tabindex="-1"></a>       <span class="op">+</span> p9.geom_col(p9.aes(fill<span class="op">=</span><span class="st">'header'</span>))</span>
<span id="cb17-15"><a aria-hidden="true" href="#cb17-15" tabindex="-1"></a>       <span class="op">+</span> p9.facet_wrap(<span class="st">'header'</span>, scales<span class="op">=</span><span class="st">'free_y'</span>, ncol<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb17-16"><a aria-hidden="true" href="#cb17-16" tabindex="-1"></a>       <span class="op">+</span> p9.coord_flip()</span>
<span id="cb17-17"><a aria-hidden="true" href="#cb17-17" tabindex="-1"></a>       <span class="op">+</span> p9.labs(x<span class="op">=</span><span class="st">''</span>, y<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb17-18"><a aria-hidden="true" href="#cb17-18" tabindex="-1"></a>       <span class="op">+</span> p9.scale_fill_manual([<span class="st">'#D85E01'</span>, <span class="st">'#7B76B5'</span>])</span>
<span id="cb17-19"><a aria-hidden="true" href="#cb17-19" tabindex="-1"></a>       <span class="op">+</span> p9.guides(fill<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb17-20"><a aria-hidden="true" href="#cb17-20" tabindex="-1"></a>       <span class="op">+</span> p9.theme(figure_size<span class="op">=</span>(<span class="dv">5</span>, <span class="dv">7</span>))</span>
<span id="cb17-21"><a aria-hidden="true" href="#cb17-21" tabindex="-1"></a>       </span>
<span id="cb17-22"><a aria-hidden="true" href="#cb17-22" tabindex="-1"></a>       )</span>
<span id="cb17-23"><a aria-hidden="true" href="#cb17-23" tabindex="-1"></a>)</span>
<span id="cb17-24"><a aria-hidden="true" href="#cb17-24" tabindex="-1"></a></span>
<span id="cb17-25"><a aria-hidden="true" href="#cb17-25" tabindex="-1"></a>f.save(<span class="st">'../output/unique_pm_species.svg'</span>, dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb17-26"><a aria-hidden="true" href="#cb17-26" tabindex="-1"></a>f.draw()</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="19">
<div>
<figure class="figure">
<p><img class="img-fluid figure-img" src="HVS-SASS_files/figure-html/cell-17-output-1.png"/></p>
</figure>
</div>
</div>
</div>
<p>If we go ahead and test whether the richness of species is different between the two head types, we can do a non-parametric paired test (Wilcoxon signed-rank test) to compare the richness of species between the two head types:</p>
<div class="cell" data-execution_count="20" id="cell-42">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a aria-hidden="true" href="#cb18-1" tabindex="-1"></a>pm_richness <span class="op">=</span> (species_long</span>
<span id="cb18-2"><a aria-hidden="true" href="#cb18-2" tabindex="-1"></a>.merge(metadata_df[[<span class="st">'Sampling'</span>, <span class="st">'Sample ID'</span>, <span class="st">'sample_id'</span>]])</span>
<span id="cb18-3"><a aria-hidden="true" href="#cb18-3" tabindex="-1"></a>.rename(columns<span class="op">=</span>{<span class="st">'Sample ID'</span>: <span class="st">'sample_code'</span>})</span>
<span id="cb18-4"><a aria-hidden="true" href="#cb18-4" tabindex="-1"></a>.query(<span class="st">'name not in @contamination_list'</span>)</span>
<span id="cb18-5"><a aria-hidden="true" href="#cb18-5" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'PM'</span>)]</span>
<span id="cb18-6"><a aria-hidden="true" href="#cb18-6" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.endswith(<span class="st">'-B'</span>)]</span>
<span id="cb18-7"><a aria-hidden="true" href="#cb18-7" tabindex="-1"></a>.loc[<span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.contains(<span class="st">'.1.'</span>, regex<span class="op">=</span><span class="va">False</span>)]</span>
<span id="cb18-8"><a aria-hidden="true" href="#cb18-8" tabindex="-1"></a>.assign(header<span class="op">=</span><span class="kw">lambda</span> dd: np.where(dd.sample_code.<span class="bu">str</span>.contains(<span class="st">"2.5"</span>, regex<span class="op">=</span><span class="va">False</span>),</span>
<span id="cb18-9"><a aria-hidden="true" href="#cb18-9" tabindex="-1"></a>                                   <span class="st">'PM2.5'</span>, <span class="st">'PM10'</span></span>
<span id="cb18-10"><a aria-hidden="true" href="#cb18-10" tabindex="-1"></a>                                   ))</span>
<span id="cb18-11"><a aria-hidden="true" href="#cb18-11" tabindex="-1"></a>.query(<span class="st">'reads &gt; 0'</span>)</span>
<span id="cb18-12"><a aria-hidden="true" href="#cb18-12" tabindex="-1"></a>.groupby([<span class="st">'header'</span>, <span class="st">'sample_code'</span>])</span>
<span id="cb18-13"><a aria-hidden="true" href="#cb18-13" tabindex="-1"></a>[<span class="st">'name'</span>]</span>
<span id="cb18-14"><a aria-hidden="true" href="#cb18-14" tabindex="-1"></a>.nunique()</span>
<span id="cb18-15"><a aria-hidden="true" href="#cb18-15" tabindex="-1"></a>.reset_index()</span>
<span id="cb18-16"><a aria-hidden="true" href="#cb18-16" tabindex="-1"></a>.assign(day<span class="op">=</span><span class="kw">lambda</span> dd: dd.sample_code.<span class="bu">str</span>.split(<span class="st">'.'</span>).<span class="bu">str</span>[<span class="op">-</span><span class="dv">1</span>])</span>
<span id="cb18-17"><a aria-hidden="true" href="#cb18-17" tabindex="-1"></a>.pivot(index<span class="op">=</span><span class="st">'day'</span>, columns<span class="op">=</span><span class="st">'header'</span>, values<span class="op">=</span><span class="st">'name'</span>)</span>
<span id="cb18-18"><a aria-hidden="true" href="#cb18-18" tabindex="-1"></a>.assign(diff<span class="op">=</span><span class="kw">lambda</span> dd: (dd[<span class="st">'PM2.5'</span>] <span class="op">-</span> dd[<span class="st">'PM10'</span>]))</span>
<span id="cb18-19"><a aria-hidden="true" href="#cb18-19" tabindex="-1"></a>)</span>
<span id="cb18-20"><a aria-hidden="true" href="#cb18-20" tabindex="-1"></a></span>
<span id="cb18-21"><a aria-hidden="true" href="#cb18-21" tabindex="-1"></a>pm_richness</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="20">
<div>
<table class="dataframe caption-top table table-sm table-striped small" data-border="1" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">header</th>
<th data-quarto-table-cell-role="th">PM10</th>
<th data-quarto-table-cell-role="th">PM2.5</th>
<th data-quarto-table-cell-role="th">diff</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">day</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">1-B</td>
<td>3682</td>
<td>3626</td>
<td>-56</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2-B</td>
<td>3019</td>
<td>3232</td>
<td>213</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3-B</td>
<td>1599</td>
<td>642</td>
<td>-957</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<p>The results of the test show a non-significant difference in the richness of species between the two head types (p-value = 0.75).</p>
<div class="cell" data-execution_count="21" id="cell-44">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a aria-hidden="true" href="#cb19-1" tabindex="-1"></a>stats.wilcoxon(pm_richness[<span class="st">'diff'</span>])</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<pre><code>WilcoxonResult(statistic=2.0, pvalue=0.75)</code></pre>
</div>
</div>
</section>
<section class="level3" id="supp.-experiment-bags-vs-vial-blanks">
<h3 class="anchored" data-anchor-id="supp.-experiment-bags-vs-vial-blanks">Supp. Experiment: Bags vs Vial blanks</h3>
<p>We are going to compare the richness of the blanks using vials and bags to see if there is a significant difference between the two types of blanks:</p>
<div class="cell" data-execution_count="22" id="cell-47">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a aria-hidden="true" href="#cb21-1" tabindex="-1"></a>bag_vial_blanks <span class="op">=</span> [<span class="st">'B6-B'</span>, <span class="st">'B6-V'</span>, <span class="st">'B3-B'</span>, <span class="st">'BPM-B'</span>, <span class="st">'B3-V'</span>, <span class="st">'BPM-V'</span>, <span class="st">'BQZ-B'</span>, <span class="st">'BQZ-V'</span>]</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="23" id="cell-48">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a aria-hidden="true" href="#cb22-1" tabindex="-1"></a>blanks_table <span class="op">=</span> pd.read_csv(<span class="st">'../data/ABS_hvs_blank2-sp.csv'</span>)</span>
<span id="cb22-2"><a aria-hidden="true" href="#cb22-2" tabindex="-1"></a>blanks_table.columns <span class="op">=</span> [col.split(<span class="st">'.'</span>)[<span class="dv">0</span>] <span class="cf">for</span> col <span class="kw">in</span> blanks_table.columns]</span>
<span id="cb22-3"><a aria-hidden="true" href="#cb22-3" tabindex="-1"></a>blanks_table.rename(columns<span class="op">=</span>{<span class="st">'#Datasets'</span>: <span class="st">'name'</span>}, inplace<span class="op">=</span><span class="va">True</span>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="24" id="cell-49">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a aria-hidden="true" href="#cb23-1" tabindex="-1"></a>blanks_richness <span class="op">=</span>(blanks_table</span>
<span id="cb23-2"><a aria-hidden="true" href="#cb23-2" tabindex="-1"></a> .melt(<span class="st">'name'</span>, var_name<span class="op">=</span><span class="st">'sample_id'</span>, value_name<span class="op">=</span><span class="st">'reads'</span>)</span>
<span id="cb23-3"><a aria-hidden="true" href="#cb23-3" tabindex="-1"></a> .merge(metadata_df[[<span class="st">'Sample ID'</span>, <span class="st">'sample_id'</span>]])</span>
<span id="cb23-4"><a aria-hidden="true" href="#cb23-4" tabindex="-1"></a> .query(<span class="st">'`Sample ID` in @bag_vial_blanks'</span>)</span>
<span id="cb23-5"><a aria-hidden="true" href="#cb23-5" tabindex="-1"></a> .assign(kind<span class="op">=</span><span class="kw">lambda</span> dd: dd[<span class="st">'Sample ID'</span>].<span class="bu">str</span>.split(<span class="st">'-'</span>).<span class="bu">str</span>[<span class="dv">1</span>]</span>
<span id="cb23-6"><a aria-hidden="true" href="#cb23-6" tabindex="-1"></a>         .replace({<span class="st">'B'</span>: <span class="st">'Bag'</span>, <span class="st">'V'</span>: <span class="st">'Vial'</span>}))</span>
<span id="cb23-7"><a aria-hidden="true" href="#cb23-7" tabindex="-1"></a> .assign(sample_name<span class="op">=</span><span class="kw">lambda</span> dd: dd[<span class="st">'Sample ID'</span>].<span class="bu">str</span>.split(<span class="st">'-'</span>).<span class="bu">str</span>[<span class="dv">0</span>])</span>
<span id="cb23-8"><a aria-hidden="true" href="#cb23-8" tabindex="-1"></a> .query(<span class="st">'reads &gt; 0'</span>)</span>
<span id="cb23-9"><a aria-hidden="true" href="#cb23-9" tabindex="-1"></a> .groupby([<span class="st">'sample_name'</span>, <span class="st">'kind'</span>])</span>
<span id="cb23-10"><a aria-hidden="true" href="#cb23-10" tabindex="-1"></a> .agg({<span class="st">'name'</span>: <span class="st">'nunique'</span>})</span>
<span id="cb23-11"><a aria-hidden="true" href="#cb23-11" tabindex="-1"></a> .reset_index()</span>
<span id="cb23-12"><a aria-hidden="true" href="#cb23-12" tabindex="-1"></a>)</span>
<span id="cb23-13"><a aria-hidden="true" href="#cb23-13" tabindex="-1"></a></span>
<span id="cb23-14"><a aria-hidden="true" href="#cb23-14" tabindex="-1"></a>(blanks_richness</span>
<span id="cb23-15"><a aria-hidden="true" href="#cb23-15" tabindex="-1"></a> .pipe(<span class="kw">lambda</span> dd: p9.ggplot(dd)</span>
<span id="cb23-16"><a aria-hidden="true" href="#cb23-16" tabindex="-1"></a>        <span class="op">+</span> p9.aes(<span class="st">'sample_name'</span>, <span class="st">'name'</span>, fill<span class="op">=</span><span class="st">'kind'</span>)</span>
<span id="cb23-17"><a aria-hidden="true" href="#cb23-17" tabindex="-1"></a>        <span class="op">+</span> p9.geom_col(position<span class="op">=</span><span class="st">'dodge'</span>, width<span class="op">=</span><span class="fl">.7</span>)</span>
<span id="cb23-18"><a aria-hidden="true" href="#cb23-18" tabindex="-1"></a>        <span class="op">+</span> p9.coord_flip()</span>
<span id="cb23-19"><a aria-hidden="true" href="#cb23-19" tabindex="-1"></a>        <span class="op">+</span> p9.labs(x<span class="op">=</span><span class="st">''</span>, y<span class="op">=</span><span class="st">'Species Richness'</span>, fill<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb23-20"><a aria-hidden="true" href="#cb23-20" tabindex="-1"></a>        <span class="op">+</span> p9.scale_fill_manual([<span class="st">'#D85E01'</span>, <span class="st">'#7B76B5'</span>])</span>
<span id="cb23-21"><a aria-hidden="true" href="#cb23-21" tabindex="-1"></a>        <span class="op">+</span> p9.theme(figure_size<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">3</span>),</span>
<span id="cb23-22"><a aria-hidden="true" href="#cb23-22" tabindex="-1"></a>                   legend_position<span class="op">=</span><span class="st">'top'</span>,</span>
<span id="cb23-23"><a aria-hidden="true" href="#cb23-23" tabindex="-1"></a>                   legend_key_size<span class="op">=</span><span class="dv">11</span>,</span>
<span id="cb23-24"><a aria-hidden="true" href="#cb23-24" tabindex="-1"></a>                   )</span>
<span id="cb23-25"><a aria-hidden="true" href="#cb23-25" tabindex="-1"></a>       )</span>
<span id="cb23-26"><a aria-hidden="true" href="#cb23-26" tabindex="-1"></a>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img class="img-fluid figure-img" src="HVS-SASS_files/figure-html/cell-22-output-1.png"/></p>
</figure>
</div>
</div>
</div>
<p>There is no significant difference when testing with a Wilcoxon signed-rank test (p-value = 0.375):</p>
</section>
<section class="level3" id="shannon-and-simpson-diversity-indexes">
<h3 class="anchored" data-anchor-id="shannon-and-simpson-diversity-indexes">Shannon and Simpson diversity indexes</h3>
<p>We are now going to compute the Shannon and Simpson diversity indexes for the blanks.</p>
<p>The Shannon diversity index H is given by the formula:</p>
<p><span class="math display">\[H = -\sum_{i=1}^{R} p_i \ln(p_i)\]</span></p>
<p>where:</p>
<ul>
<li><span class="math inline">\(R\)</span> is the total number of species,</li>
<li><span class="math inline">\(p_i\)</span> is the proportion of individuals belonging to the <span class="math inline">\(i\)</span>-th species, calculated as <span class="math inline">\(p_i = \frac{n_i}{N}\)</span> ,</li>
<li><span class="math inline">\(n_i\)</span> is the number of individuals in the i -th species,</li>
<li><span class="math inline">\(N\)</span> is the total number of individuals across all species.</li>
</ul>
<p>The Simpson diversity index D can then be defined as:</p>
<p><span class="math display">\[D = 1 - \sum_{i=1}^{R} p_i^2  = 1 - \frac{\sum_{i=1}^{R} n_i (n_i - 1)}{N (N - 1)} \]</span></p>
<p>We’ll define a function to calculate each of these indexes in numpy and then simply apply it to the columns of the dataframe with the species counts to get the diversity indexes for each sample:</p>
<div class="cell" data-execution_count="74" id="cell-53">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a aria-hidden="true" href="#cb24-1" tabindex="-1"></a><span class="im">from</span> typing <span class="im">import</span> List</span>
<span id="cb24-2"><a aria-hidden="true" href="#cb24-2" tabindex="-1"></a></span>
<span id="cb24-3"><a aria-hidden="true" href="#cb24-3" tabindex="-1"></a><span class="kw">def</span> shannon_diversity(species_counts: List[<span class="bu">int</span>]) <span class="op">-&gt;</span> <span class="bu">float</span>:</span>
<span id="cb24-4"><a aria-hidden="true" href="#cb24-4" tabindex="-1"></a>    species_counts <span class="op">=</span> np.array(species_counts)</span>
<span id="cb24-5"><a aria-hidden="true" href="#cb24-5" tabindex="-1"></a>    species_counts <span class="op">=</span> species_counts[species_counts <span class="op">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb24-6"><a aria-hidden="true" href="#cb24-6" tabindex="-1"></a>    total_counts <span class="op">=</span> np.<span class="bu">sum</span>(species_counts)</span>
<span id="cb24-7"><a aria-hidden="true" href="#cb24-7" tabindex="-1"></a>    proportions <span class="op">=</span> species_counts <span class="op">/</span> total_counts</span>
<span id="cb24-8"><a aria-hidden="true" href="#cb24-8" tabindex="-1"></a>    shannon_index <span class="op">=</span> <span class="op">-</span>np.<span class="bu">sum</span>(proportions <span class="op">*</span> np.log(proportions))</span>
<span id="cb24-9"><a aria-hidden="true" href="#cb24-9" tabindex="-1"></a>    <span class="cf">return</span> shannon_index</span>
<span id="cb24-10"><a aria-hidden="true" href="#cb24-10" tabindex="-1"></a></span>
<span id="cb24-11"><a aria-hidden="true" href="#cb24-11" tabindex="-1"></a><span class="kw">def</span> simpson_diversity(species_counts: List[<span class="bu">int</span>]) <span class="op">-&gt;</span> <span class="bu">float</span>:</span>
<span id="cb24-12"><a aria-hidden="true" href="#cb24-12" tabindex="-1"></a>    species_counts <span class="op">=</span> np.array(species_counts)</span>
<span id="cb24-13"><a aria-hidden="true" href="#cb24-13" tabindex="-1"></a>    total_counts <span class="op">=</span> np.<span class="bu">sum</span>(species_counts)</span>
<span id="cb24-14"><a aria-hidden="true" href="#cb24-14" tabindex="-1"></a>    proportions <span class="op">=</span> species_counts <span class="op">/</span> total_counts</span>
<span id="cb24-15"><a aria-hidden="true" href="#cb24-15" tabindex="-1"></a>    simpson_index <span class="op">=</span> <span class="dv">1</span> <span class="op">-</span> np.<span class="bu">sum</span>(proportions <span class="op">**</span> <span class="dv">2</span>)</span>
<span id="cb24-16"><a aria-hidden="true" href="#cb24-16" tabindex="-1"></a>    <span class="cf">return</span> simpson_index</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="101" id="cell-54">
<details class="code-fold">
<summary>Show Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a aria-hidden="true" href="#cb25-1" tabindex="-1"></a>blank_sample_names <span class="op">=</span> metadata_df.loc[metadata_df[<span class="st">'Sample ID'</span>].<span class="bu">str</span>.startswith(<span class="st">'B'</span>)][<span class="st">'Sample ID'</span>]</span>
<span id="cb25-2"><a aria-hidden="true" href="#cb25-2" tabindex="-1"></a></span>
<span id="cb25-3"><a aria-hidden="true" href="#cb25-3" tabindex="-1"></a>(blanks_table</span>
<span id="cb25-4"><a aria-hidden="true" href="#cb25-4" tabindex="-1"></a> .melt(<span class="st">'name'</span>, var_name<span class="op">=</span><span class="st">'sample_id'</span>, value_name<span class="op">=</span><span class="st">'counts'</span>)</span>
<span id="cb25-5"><a aria-hidden="true" href="#cb25-5" tabindex="-1"></a> .merge(metadata_df[[<span class="st">'Sample ID'</span>, <span class="st">'sample_id'</span>]])</span>
<span id="cb25-6"><a aria-hidden="true" href="#cb25-6" tabindex="-1"></a> .query(<span class="st">'`Sample ID` in @blank_sample_names'</span>)</span>
<span id="cb25-7"><a aria-hidden="true" href="#cb25-7" tabindex="-1"></a> .pivot(index<span class="op">=</span><span class="st">'name'</span>, columns<span class="op">=</span><span class="st">'Sample ID'</span>, values<span class="op">=</span><span class="st">'counts'</span>)</span>
<span id="cb25-8"><a aria-hidden="true" href="#cb25-8" tabindex="-1"></a> .fillna(<span class="dv">0</span>)</span>
<span id="cb25-9"><a aria-hidden="true" href="#cb25-9" tabindex="-1"></a> .pipe(<span class="kw">lambda</span> dd: pd.concat([</span>
<span id="cb25-10"><a aria-hidden="true" href="#cb25-10" tabindex="-1"></a>        dd.<span class="bu">apply</span>(shannon_diversity, axis<span class="op">=</span><span class="dv">0</span>).rename(<span class="st">'Shannon Index'</span>).reset_index()</span>
<span id="cb25-11"><a aria-hidden="true" href="#cb25-11" tabindex="-1"></a>        .merge(dd.<span class="bu">apply</span>(simpson_diversity, axis<span class="op">=</span><span class="dv">0</span>).rename(<span class="st">'Simpson Index'</span>).reset_index())</span>
<span id="cb25-12"><a aria-hidden="true" href="#cb25-12" tabindex="-1"></a>    ])</span>
<span id="cb25-13"><a aria-hidden="true" href="#cb25-13" tabindex="-1"></a> )</span>
<span id="cb25-14"><a aria-hidden="true" href="#cb25-14" tabindex="-1"></a> .melt(<span class="st">'Sample ID'</span>, var_name<span class="op">=</span><span class="st">'index'</span>, value_name<span class="op">=</span><span class="st">'value'</span>)</span>
<span id="cb25-15"><a aria-hidden="true" href="#cb25-15" tabindex="-1"></a> .pipe(<span class="kw">lambda</span> dd: p9.ggplot(dd)</span>
<span id="cb25-16"><a aria-hidden="true" href="#cb25-16" tabindex="-1"></a>       </span>
<span id="cb25-17"><a aria-hidden="true" href="#cb25-17" tabindex="-1"></a>       <span class="op">+</span> p9.aes(<span class="st">'Sample ID'</span>, <span class="st">'value'</span>)</span>
<span id="cb25-18"><a aria-hidden="true" href="#cb25-18" tabindex="-1"></a>       <span class="op">+</span> p9.geom_col()</span>
<span id="cb25-19"><a aria-hidden="true" href="#cb25-19" tabindex="-1"></a>       <span class="op">+</span> p9.facet_wrap(<span class="st">'~index'</span>, scales<span class="op">=</span><span class="st">'free_y'</span>, ncol<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb25-20"><a aria-hidden="true" href="#cb25-20" tabindex="-1"></a>       <span class="op">+</span> p9.labs(x<span class="op">=</span><span class="st">''</span>, y<span class="op">=</span><span class="st">''</span>)</span>
<span id="cb25-21"><a aria-hidden="true" href="#cb25-21" tabindex="-1"></a>       <span class="op">+</span> p9.theme(figure_size<span class="op">=</span>(<span class="dv">4</span>, <span class="dv">3</span>), axis_text_x<span class="op">=</span>p9.element_text(angle<span class="op">=</span><span class="dv">45</span>))</span>
<span id="cb25-22"><a aria-hidden="true" href="#cb25-22" tabindex="-1"></a> )</span>
<span id="cb25-23"><a aria-hidden="true" href="#cb25-23" tabindex="-1"></a>)</span></code><button class="code-copy-button" title="Copy to Clipboard"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img class="img-fluid figure-img" src="HVS-SASS_files/figure-html/cell-24-output-1.png"/></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<script src="site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>